---
title: "MVP Extended Data"
author: "Kyle Inman"
date: "`r Sys.Date()`"
output:
  pdf_document:
    pandoc_args: --listings
    includes:
      in_header: "header.tex"
    toc: true
    toc_depth: 4
    number_sections: true
    extra_dependencies: ["float"]
    
  word_document: default
  html_document: default
---

# Setup
```{r Setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(tidy = TRUE, tidy.opts = list(width.cutoff = 100))
knitr::opts_chunk$set(out.width = "90%")

library(pagedown)
library(namer)
library(tidyverse)
library(formatR)
library(ff)
library(LDlinkR)
library(corrplot)
library(robustbase)
library(webshot)
library(readxl)
library(ggplot2)
library(ggcorrplot)
library(Hmisc)
library(psych)
library(lubridate)
library(stargazer)
library(reshape2)
library(ggbeeswarm)
library(epitools)
library(vegan)
library(tidyr)
library(gdata)
library(stringr)
library(ggpubr)
library(mice)
library(data.table)
library(survival)
library(readr)
library("Partiallyoverlapping")
library("haven")
library(tableone)
library(RColorBrewer)
library(pscl)
library(survminer)
library(ggsurvfit)
library(purrr)
library(tableone)
library(gtsummary)
library(adjustedCurves)
library(dplyr)
library(ggrepel)
library(LDlinkR)
library(webshot2)
library(conflicted)
library(tinytex)

ld_key <- Sys.getenv('LDLINK_TOKEN')

conflicts_prefer(dplyr::mutate)
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::select)
conflicts_prefer(gdata::first)
conflicts_prefer(dplyr::summarize)
conflicts_prefer(dplyr::rename)
conflicts_prefer(plyr::count)
conflicts_prefer(dplyr::arrange)
conflicts_prefer(base::rbind)
conflicts_prefer(dplyr::slice)
conflicts_prefer(webshot::webshot)
conflicts_prefer(gdata::starts_with)
conflicts_prefer(gt::pct)
```

# Import gene positions and filtered version of SNP database (only alternative pathway and sepsis Phecodes) - filtered in command line due to inability to import into RStudio (because of size). Set working directory, split into different racial groups, create complement groupings and select genes/conditions of interest
```{r Data import and labeling}
setwd("/Users/kinman/Desktop/MVP_Data/")
snps_extended <- read.table("snps_altsepsis.txt", sep="\t", na.string="99", header=TRUE)
n_distinct(snps_extended$rsid)

#Import dataframe containing genes with GRCh38 start/end positions
genes_by_loci <- read.table("MVP_genes_loci_AP.txt", sep="\t", na.string="99", header=TRUE)
genes_by_loci$Chromosome <- as.character(genes_by_loci$Chromosome)
genes_by_loci$Start <- as.numeric(genes_by_loci$Start)
genes_by_loci$End <- as.numeric(genes_by_loci$End)

#Label complement groups
snps_extended <- snps_extended %>% mutate(complement_group=case_when(
  Gene=="CFD" | Gene=="CFB" | Gene=="CFP" | Gene=="CFI" | Gene=="CFH" | Gene=="C3" ~ "alternative",
  Gene=="C1QA" | Gene=="C1QB" | Gene=="C1QC" | Gene=="C1R" | Gene=="C1S" | Gene=="C2" | Gene=="C4A" | Gene=="C4B" | Gene=="C4BPA" | Gene=="C4BPB" ~ "classical",
  Gene=="MBL2" | Gene=="MASP1" | Gene=="MASP2" ~ "lectin",
  Gene=="CR1" | Gene=="CR2" | Gene=="C3AR1" | Gene=="C5AR1" | Gene=="C5AR2" ~ "receptor",
  Gene=="CD46" | Gene=="CD55" | Gene=="CD59" ~ "regulator",
  Gene=="C5" | Gene=="C6" | Gene=="C7" | Gene=="C8A" | Gene=="C8B" | Gene=="C8G" | Gene=="C9" ~ "terminal",
  Gene=="CFHR1" | Gene=="CFHR2" | Gene=="CFHR3" | Gene=="CFHR4" ~ "other"))

#Label Phecodes
snps_extended <- snps_extended %>% mutate(condition=case_when(
  Phecode=="Phe_480" ~ "Pneumonia",
  Phecode=="Phe_480_1" ~ "Bacterial pneumonia",
  Phecode=="Phe_480_2" ~ "Viral pneumonia",
  Phecode=="Phe_480_3" ~ "Pneumonia due to fungus (mycoses)",
  Phecode=="Phe_480_5" ~ "Bronchopneumonia and lung abscess",
  Phecode=="Phe_480_11" ~ "Pneumococcal pneumonia",
  Phecode=="Phe_481" ~ "Influenza",
  Phecode=="Phe_483" ~ "Acute bronchitis and bronchiolitis",
  Phecode=="Phe_038" ~ "Septicemia",
  Phecode=="Phe_038_1" ~ "Gram negative septicemia",
  Phecode=="Phe_038_2" ~ "Gram positive septicemia",
  Phecode=="Phe_038_3" ~ "Bacteremia",
  Phecode=="Phe_117" ~ "Mycoses",
  Phecode=="Phe_994" ~ "Sepsis and SIRS",
  Phecode=="Phe_994_2" ~ "Sepsis",
  Phecode=="Phe_994_21" ~ "Septic shock"))

Phecodes_of_interest <- c("Phe_480", "Phe_480_1", "Phe_480_2", "Phe_480_3", "Phe_480_5", "Phe_480_11", "Phe_481", "Phe_483", "Phe_038", "Phe_038_1", "Phe_038_2", "Phe_038_3", "Phe_117", "Phe_994", "Phe_994_2", "Phe_994_21")

conditions_of_interest <- c("Pneumonia", "Bacterial pneumonia", "Viral pneumonia", "Pneumonia due to fungus (mycoses)", "Bronchopneumonia and lung abscess", "Pneumococcal pneumonia", "Influenza", "Acute bronchitis and bronchiolitis", "Septicemia", "Gram negative septicemia", "Gram positive septicemia", "Bacteremia", "Mycoses", "Sepsis and SIRS", "Sepsis", "Septic shock")

combined_phecodes_conditions <- data.frame(Phecode=Phecodes_of_interest, condition=conditions_of_interest)

snps_meta_ext <- filter(snps_extended, Race=="META")
snps_afr_ext <- filter(snps_extended, Race=="AFR")
snps_eur_ext <- filter(snps_extended, Race=="EUR")
snps_his_ext <- filter(snps_extended, Race=="HIS")
#snps_asn <- filter(snps_extended, Race=="ASN")
#Not enough Asian data

#Pull all the sepsis and bacteremia conditions
sepsis_conditions <- c("Septicemia", "Gram negative septicemia", "Gram positive septicemia", "Bacteremia", "Sepsis and SIRS", "Sepsis", "Septic shock")

#Break down into bacteremia phenotypes (infection) and sepsis phenotypes (inflammation)
infection_conditions <- c("Septicemia", "Gram negative septicemia", "Gram positive septicemia", "Bacteremia")
inflammation_conditions <- c("Sepsis and SIRS", "Sepsis", "Septic shock")

genes_of_interest <- c("CFD", "CFB", "CFP", "CFI", "CFH", "C3", "CR1", "CR2", "C3AR1", "C5AR1", "C5AR2", "CFHR1", "CFHR2", "CFHR3", "CFHR4", "CD46", "CD55", "CD59")

n_distinct(snps_meta_ext$rsid)
```

# Add case and control MAFs to the snps_extended data frame
```{r Adding case and control MAFs}
snps_extended <- snps_extended %>% mutate(case_maf = (1-case_af))
snps_extended <- snps_extended %>% mutate(control_maf = (1-control_af))
```

# Filter to only significant SNPs with MAF >0.01
```{r Filtering to significant and common}
psig = 0.05
maf_cutoff = 0.01

snps_altsig_filt <- snps_extended %>% filter(pval<psig & maf>maf_cutoff & Gene %in% genes_of_interest & condition %in% sepsis_conditions)
n_distinct(snps_altsig_filt$rsid)
```

# Labeling the SNPs based on the gene they are in for the purposes of significance filtering
```{r Labeling based on gene location}
#Take the first gene that matches and label SNP that way and create a variable to easily identify genic SNPs
snps_altsig_filt$chrom <- as.character(snps_altsig_filt$chrom)

snps_altsig_filt <- snps_altsig_filt %>%
  rowwise() %>%
  mutate(
    Gene = case_when(
      any(pos >= genes_by_loci$Start & pos <= genes_by_loci$End) ~ 
        genes_by_loci$Gene[which(pos >= genes_by_loci$Start & pos <= genes_by_loci$End)][1],
      TRUE ~ Gene
    ),
    in_gene = case_when(any(pos >= genes_by_loci$Start & pos <= genes_by_loci$End) ~ 1,
                        TRUE ~ 0)) %>% 
  ungroup()
```

# Import the Li and Ji results (previously calculated) and add gene-specific thresholds to identify the significant SNPs with p-value correction based on their gene annotations
```{r Importing Li and Ji results}
lj_results <- read.table("lj_results_v2.txt", sep = "\t", header=TRUE)

snps_altsig_filt_temp <- snps_altsig_filt %>% left_join(., (lj_results %>% select("sig_threshold_bon","Gene")), by="Gene")
snps_altsig_filt <- snps_altsig_filt_temp %>% mutate(lj_sig = case_when(pval < sig_threshold_bon ~ 1, TRUE ~ 0))
snps_altsig_filt$lj_sig <- as.numeric(snps_altsig_filt$lj_sig)
n_distinct(snps_altsig_filt$rsid)
```

# Removing duplicate entries (ie, as a SNP and an indel)
```{r Removing duplicates}
snps_noindels_altsig_filt <- snps_altsig_filt %>% group_by(Phecode, Race, Gene, rsid) %>% arrange(ID) %>% slice(1)
n_distinct(snps_noindels_altsig_filt$rsid)
```

# Make a dataframe containing race, condition, and number of cases to add for FUMA annotation (the program wants a num_cases column though this does not affect the output)
```{r Making dataframe with race, condition, and num_cases for later use}
race_condition_numcases <- snps_noindels_altsig_filt %>% group_by(rsid) %>% mutate(num_cases_sum = sum(num_cases[Race %in% c("AFR", "EUR", "HIS")], na.rm = TRUE), num_cases = ifelse(Race == "META", num_cases_sum, num_cases)) %>% ungroup() %>% select(Race, condition, num_cases) %>% group_by(Race, condition) %>% slice(1)
```

# Create dataframe to convert between GRCh37 and GRCh38 positions
```{r Creating GRCh37 and GRCh38 conversion dataframe}
rsid_pos <- snps_noindels_altsig_filt %>% ungroup() %>% select(rsid, pos) %>% rename("pos_grch38" = "pos") %>% group_by(rsid) %>% slice(1)
```

# Add number of cases and then prepare for export to FUMA
```{r MVP-ExtendedData-Final-v1.7-9}
#Fill in missing num_cases for META groups -> actual value doesn't matter in our analysis, but FUMA requires this for uploaded data
snps_noindels_altsig_filt_numcases <- snps_noindels_altsig_filt %>% 
  group_by(rsid) %>% 
  mutate(num_cases_sum = sum(num_cases[Race %in% c("AFR", "EUR", "HIS")], na.rm = TRUE), 
         num_cases = ifelse(Race == "META", num_cases_sum, num_cases)) %>% 
  ungroup()

#Get list of significant SNPs in sepsis samples within the alternative/receptor groupings
snps_sepsis_ext <- snps_noindels_altsig_filt_numcases %>% filter(lj_sig==1 & Gene %in% genes_of_interest & maf > maf_cutoff & condition %in% sepsis_conditions) %>% arrange(rsid)
n_distinct(snps_sepsis_ext)

#Convert SNPs to GRCh37 for FUMA, then select only the columns we want for FUMA
snps_sepsis_ext_grch37 <- snps_sepsis_ext
snps_sepsis_ext_grch37$pos <- sapply(strsplit(as.character(snps_sepsis_ext$ID), ":"), `[`, 2)
fuma_snps_ext <- snps_sepsis_ext_grch37 %>% select(rsid,chrom,pos,ref,alt,beta,sebeta,or,pval,num_cases)

#Remove NA p-values and make p-value smaller so it is usable by FUMA -> NOT using this version for actual statistical analysis, just FUMA annotations!
fuma_snps_ext <- fuma_snps_ext %>% filter(!is.na(pval))
fuma_snps_ext_pvalinc <- fuma_snps_ext %>% rowwise() %>% mutate(pval = pval/100000) %>% filter(!is.na(alt) & !is.na(ref) & !is.na(pval)) %>% group_by(rsid) %>% slice(1) %>% ungroup() %>% na.omit() #Making all p-values reach FUMA's stringent cutoff since this is not a GWAS
fuma_snps_ext_pvalinc$num_cases[fuma_snps_ext_pvalinc$num_cases==0] <- 637 #Num_cases doesn't actually affect output, but setting 0 to lowest num_cases value so SNPs aren't completely skipped by FUMA just in case
n_distinct(fuma_snps_ext_pvalinc$rsid)

#Export the updated FUMA dataframe
write.table(fuma_snps_ext_pvalinc, "fuma_snps_ext_pvalinc.txt", sep="\t", quote=FALSE, row.names=FALSE)
```

# Importing and annotating using FUMA data
```{r MVP-ExtendedData-Final-v1.7-10}
fuma_data_ext_snps <- read.table(file="FUMA_job600066/snps.txt", sep="\t", na.string="NA", header=TRUE)
fuma_data_ext_genes <- read.table(file="FUMA_job600066/genes.txt", sep="\t", na.string="NA", header=TRUE)
fuma_data_ext_annov <- read.table(file="FUMA_job600066/annov.txt", sep="\t", na.string="NA", header=TRUE)
fuma_data_ext_annot <- read.table(file="FUMA_job600066/annot.txt", sep="\t", na.string="NA", header=TRUE)
fuma_data_ext_eqtl <- read.table(file="FUMA_job600066/eqtl.txt", sep="\t", na.string="NA", header=TRUE)
fuma_data_ext_ciprom <- read.table(file="FUMA_job600066/ciProm.txt", sep="\t", na.string="NA", header=TRUE)
fuma_data_ext_cisnp <- read.table(file="FUMA_job600066/ciSNPs.txt", sep="\t", na.string="NA", header=TRUE)

#Creating a uniqID to rsid conversion dataframe to use rsid for all of these
uniqID_rsid_conversion <- fuma_data_ext_snps %>% select(uniqID, rsID)
names(uniqID_rsid_conversion)[names(uniqID_rsid_conversion)=="rsID"] <- "rsid"

#Prepare for merging
snps_sepsis_ext_grch37$pos <- as.numeric(snps_sepsis_ext_grch37$pos)
fuma_data_ext_snps$chr <- as.character(fuma_data_ext_snps$chr)
fuma_data_ext_annov$chr <- as.character(fuma_data_ext_annov$chr)
fuma_data_ext_genes$chr <- as.character(fuma_data_ext_genes$chr)
fuma_data_ext_eqtl$chr <- as.character(fuma_data_ext_eqtl$chr)
fuma_data_ext_cisnp$chr <- as.character(fuma_data_ext_cisnp$chr)

#Filter ANNOVAR to the genes we want and prioritize the impactful associations
#First see what labels we have
table(fuma_data_ext_annov$annot)
fuma_data_ext_annov_rsid <- fuma_data_ext_annov %>% left_join(., uniqID_rsid_conversion, by="uniqID")

#Then rank these - will do exonic > UTR3 (site for post-translational modifications) > UTR5 (site for transcription factor binding) > intronic > intergenic > other (ncRNA and splicing variants)
names(fuma_data_ext_annov_rsid)[names(fuma_data_ext_annov_rsid)=="dist"] <- "dist_annov"
fuma_data_ext_annov_rsid <- fuma_data_ext_annov_rsid %>% rowwise() %>% mutate(impact_rank = case_when(
  annot == "exonic" ~ 1,
  annot == "UTR3" | annot == "downstream" ~ 2,
  annot == "UTR5" | annot == "upstream" ~ 2,
  annot == "intronic" ~ 3,
  annot == "intergenic" ~ 4,
  TRUE ~ 5
  )) %>% ungroup()

#Filter to only the most impactful (ie, lowest rank)
fuma_data_ext_annov_filt <- fuma_data_ext_annov_rsid %>% group_by(rsid) %>% mutate(greatest_impact = case_when(impact_rank == min(impact_rank) ~ 1,
            TRUE ~ 0)) %>% ungroup()
fuma_data_ext_annov_filt <- fuma_data_ext_annov_filt %>% filter(greatest_impact == 1)

#Combining with SNP annotation
fuma_snps_combined <- snps_sepsis_ext_grch37 %>% left_join(., (fuma_data_ext_snps %>% select(-beta, -se, -or, -r2, -uniqID)), by=c("rsid"="rsID", "chrom"="chr", "pos"))
fuma_snps_combined$dist[fuma_snps_combined$dist=="0:0"] <- 0
fuma_snps_combined$dist <- as.numeric(fuma_snps_combined$dist)
fuma_snps_combined$dist[is.na(fuma_snps_combined$dist)] <- 0

#Add ANNOVAR annotations
fuma_snps_anno_combined <- fuma_snps_combined %>% left_join(., (fuma_data_ext_annov_filt %>% select(-uniqID)), by=c("chrom"="chr", "pos", "rsid"), relationship = "many-to-many") %>% arrange(rsid)

#Add gene-level annotation
fuma_snps_anno_gene_combined <- (fuma_snps_anno_combined %>% select(-GenomicLocus)) %>% left_join(., fuma_data_ext_genes, by=c("gene"="ensg", "chrom"="chr", "symbol"))
names(fuma_snps_anno_gene_combined)[names(fuma_snps_anno_gene_combined) == "type"] <- "gene_type"

#Add eQTL data and then rename columns to distinguish the gene the SNP is near from the gene it affects based on eQTL data
fuma_snps_anno_gene_eqtl_combined <- fuma_snps_anno_gene_combined %>% left_join(., (fuma_data_ext_eqtl %>% select(-uniqID)), by = c("chrom"="chr", "pos", "eqtlMapFilt"), relationship = "many-to-many")
names(fuma_snps_anno_gene_eqtl_combined)[names(fuma_snps_anno_gene_eqtl_combined) == "symbol.y"] <- "eqtl_symbol"
names(fuma_snps_anno_gene_eqtl_combined)[names(fuma_snps_anno_gene_eqtl_combined) == "symbol.x"] <- "anno_symbol"
names(fuma_snps_anno_gene_eqtl_combined)[names(fuma_snps_anno_gene_eqtl_combined) == "gene.x"] <- "anno_gene"
names(fuma_snps_anno_gene_eqtl_combined)[names(fuma_snps_anno_gene_eqtl_combined) == "gene.y"] <- "eqtl_gene"

#Condense and then add enhancer region info
fuma_data_ext_cisnp_collapsed <- fuma_data_ext_cisnp %>% group_by(uniqID, rsID, chr, pos, reg_region, type) %>% summarize(enh_tissue_cell_all = paste(na.omit(unique(tissue.cell)), collapse = ", "), .groups = 'drop') %>% ungroup()
names(fuma_data_ext_cisnp_collapsed)[names(fuma_data_ext_cisnp_collapsed)=="reg_region"] <- "enh_reg_region"

fuma_snps_anno_gene_eqtl_ci_combined <- fuma_snps_anno_gene_eqtl_combined %>% left_join(., na.omit((fuma_data_ext_cisnp_collapsed %>% select(-uniqID))), by=c("rsid"="rsID", "chrom"="chr", "pos"))
names(fuma_snps_anno_gene_eqtl_ci_combined)[names(fuma_snps_anno_gene_eqtl_ci_combined) == "type"] <- "type_enh"

#Condense and then add promoter region info
fuma_data_ext_ciprom_collapsed <- fuma_data_ext_ciprom %>% group_by(genes, reg_region, region2, type) %>% summarize(prom_tissue_cell_all = paste(na.omit(unique(tissue.cell)), collapse = ", "), .groups = 'drop') %>% ungroup() %>% na.omit(genes)
names(fuma_data_ext_ciprom_collapsed)[names(fuma_data_ext_ciprom_collapsed)=="reg_region"] <- "prom_reg_region"

fuma_snps_anno_gene_eqtl_ci2_combined <- fuma_snps_anno_gene_eqtl_ci_combined %>% left_join(., fuma_data_ext_ciprom_collapsed, by=c("anno_gene"="genes"), relationship = "many-to-many") %>% arrange(rsid)
names(fuma_snps_anno_gene_eqtl_ci2_combined)[names(fuma_snps_anno_gene_eqtl_ci2_combined) == "type"] <- "type_prom"

#Filtering to chromatin interactions only in the cells we're interested in
ci_tissues_ofinterest <- c("E029", "E030", "E032", "E034", "E037", "E038", "E039", "E040", "E041", "E042", "E043", "E044", "E045", "E046", "E047", "E048", "E050", "E051", "E062", "E066", "E017", "E088", "E096", "E113")
fuma_data_ext_annot_rsid <- fuma_data_ext_annot %>% left_join(., uniqID_rsid_conversion, by="uniqID")
names(fuma_data_ext_annot_rsid)[names(fuma_data_ext_annot_rsid)=="CADD"] <- "ci_CADD"
names(fuma_data_ext_annot_rsid)[names(fuma_data_ext_annot_rsid)=="RDB"] <- "ci_RDB"

fuma_data_ext_all <- fuma_snps_anno_gene_eqtl_ci2_combined %>% left_join(., na.omit((fuma_data_ext_annot_rsid %>% select(rsid, ci_CADD, ci_RDB, all_of(ci_tissues_ofinterest)))), by="rsid") 

#Make numeric versions
fuma_data_ext_all$CADD <- as.numeric(fuma_data_ext_all$CADD)
fuma_data_ext_all$ci_CADD <- as.numeric(fuma_data_ext_all$ci_CADD)
fuma_data_ext_all <- fuma_data_ext_all %>% mutate(CADD = if_else(is.na(CADD), ci_CADD, CADD))
fuma_data_ext_all <- fuma_data_ext_all %>% mutate(RDB = if_else(is.na(RDB), ci_RDB, RDB))

fuma_data_ext_all <- fuma_data_ext_all %>% mutate(RDB_num = case_when(RDB == "1a" | RDB == "1b" | RDB == "1c" | RDB == "1d" | RDB == "1e" | RDB == "1f" ~ "1",
                                                                         RDB == "2a" | RDB == "2b" | RDB == "2c" ~ "2",
                                                                         RDB == "3a" | RDB == "3b" ~ "3",
                                                                         RDB == "4" ~ "4",
                                                                         RDB == "5" ~ "5",
                                                                         RDB == "6" ~ "6",
                                                                         RDB == "7" ~ "7"
                                                                         ))
fuma_data_ext_all$RDB_num <- as.numeric(fuma_data_ext_all$RDB_num)
n_distinct(fuma_data_ext_all$rsid)
```

# Missing gene annotations - export to VEP for a different annotation approach
```{r MVP-ExtendedData-Final-v1.7-11}
#Counting SNPs that are completely missing Gene data - 483 cases across 193 SNPs
table(is.na(fuma_data_ext_all$anno_symbol) & is.na(fuma_data_ext_all$nearestGene))
n_distinct(fuma_data_ext_all$rsid[is.na(fuma_data_ext_all$anno_symbol) & is.na(fuma_data_ext_all$nearestGene)])

#Exporting list for manual annotation with VEP
missing_gene_list <- fuma_data_ext_all %>% filter(is.na(anno_symbol) & is.na(nearestGene)) %>% distinct(rsid) %>% unlist()
write.table(missing_gene_list, "missing_gene_snps_list.txt", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

# Import and filter the VEP annotations that we obtained from the online VEP tool to add to our dataframe
```{r MVP-ExtendedData-Final-v1.7-12}
#Importing VEP results
vep_gene_results <- read.table(
    "VEP_missing_gene_results.txt",
    header = TRUE,
    sep = "\t",
    quote = "",
    fill = TRUE,
    comment.char = "",
    stringsAsFactors = FALSE,
    na.strings = "-"   # Marks "-" as NA since this is how VEP reports things
)

#Filtering the VEP results
table(vep_gene_results$BIOTYPE)
vep_gene_filt <- vep_gene_results %>% filter(CANONICAL == "YES" | SYMBOL %in% genes_of_interest)
table(missing_gene_list %in% vep_gene_filt$Uploaded_variation)
table(vep_gene_filt$Consequence)

#Arrange by severity of impact
vep_gene_labeled <- vep_gene_filt %>% 
  mutate(
    var_impact = case_when(
      Consequence %in% c("missense_variant", "frameshift_variant", "inframe_insertion") ~ 1,
      str_detect(Consequence, "synonymous_variant") ~ 2,
      Consequence %in% c("upstream_gene_variant", "downstream_gene_variant", "3_prime_UTR_variant", "5_prime_UTR_variant") ~ 3,
      str_detect(Consequence, "intron_variant") ~ 4,
      TRUE ~ 5
    )
  )

#Create a "greatest impact" variable for each SNP
vep_gene_labeled <- vep_gene_labeled %>% group_by(Uploaded_variation) %>% mutate(
  greatest_impact = case_when(
    var_impact == min(var_impact) ~ 1,
    TRUE ~ 0
  )
)

#Filter to the greatest impact transcripts or ones in our candidate genes, then cut out any duplicate gene entries and prioritize the transcripts that are protein_coding, have associated gene names, and have the greatest impact
vep_gene_priority <- vep_gene_labeled %>% 
  filter((SYMBOL %in% genes_of_interest & SYMBOL != "C1RL" & SYMBOL != "CCHCR1") | greatest_impact == 1) %>% 
  group_by(Uploaded_variation) %>% 
  mutate(gene_count = n_distinct(SYMBOL)) %>% 
  ungroup() %>% 
  group_by(Uploaded_variation, SYMBOL) %>% 
  arrange(desc(BIOTYPE=="protein_coding"), desc(!is.na(SYMBOL)), var_impact) %>% 
  slice(1) %>% 
  arrange(Uploaded_variation)

#Filter to just 1 gene per SNP, prioritize protein_coding and genes of interest
vep_gene_final <- vep_gene_priority %>% group_by(Uploaded_variation) %>% arrange(desc(SYMBOL %in% genes_of_interest), desc(!is.na(SYMBOL)), desc(BIOTYPE=="protein_coding"), desc(grepl("RNA", BIOTYPE))) %>% slice(1) %>% ungroup()

#Simplify for addition to our dataframe
vep_gene_simple <- vep_gene_final %>% select(Uploaded_variation, SYMBOL) %>% rename("VEP_Gene" = "SYMBOL", "rsid" = "Uploaded_variation")

#Checking how many variants are missing data despite all of this -> 6 are missing
table(missing_gene_list %in% vep_gene_simple$rsid)

#Collect data on the missing SNPs and filter to one Gene each -> almost all are NA
missing_snps <- fuma_data_ext_all %>% filter(rsid %in% missing_gene_list & !(rsid %in% vep_gene_filt$Uploaded_variation))
missing_snps_vep <- vep_gene_results %>% filter(Uploaded_variation %in% missing_snps$rsid)
missing_snps_simple <- missing_snps_vep %>% group_by(Uploaded_variation) %>% arrange(desc(!is.na(SYMBOL))) %>% slice(1) %>% select(Uploaded_variation, SYMBOL) %>% rename("VEP_Gene" = "SYMBOL", "rsid" = "Uploaded_variation")

#Combine missing data with other VEP dataframe and then combine with our overall dataframe
vep_gene_simple_combined <- rbind(vep_gene_simple, missing_snps_simple)
fuma_data_ext_all_withvep <- fuma_data_ext_all %>% left_join(vep_gene_simple_combined, by="rsid")
n_distinct(fuma_data_ext_all_withvep$rsid)

#Confirm that this now contains all of the rsIDs that were missing gene annotations - it does
table(missing_gene_list %in% fuma_data_ext_all_withvep$rsid)
```

# Correct the Gene label based on our annotations
```{r MVP-ExtendedData-Final-v1.7-13}
#Label by the ANNOVAR symbol followed by the nearest gene followed by VEP_Gene. If none of those have any values, mark with a question mark to identify those where the gene association is completely unknown
fuma_data_ext_genescorrected <- fuma_data_ext_all_withvep %>% mutate(Gene = if_else(!is.na(anno_symbol), anno_symbol, if_else(!is.na(nearestGene), nearestGene, if_else(!is.na(VEP_Gene), VEP_Gene, "?"))))

fuma_data_ext_final <- fuma_data_ext_genescorrected %>% separate_longer_delim(Gene, delim = ":") %>% separate_longer_delim(nearestGene, delim = ":") %>% unnest(cols = c(Gene, nearestGene))
n_distinct(fuma_data_ext_final$rsid)

#See how many are missing gene annotations - 82 total instances of 36 unique SNPs
table(fuma_data_ext_final$Gene == "?")
n_distinct(fuma_data_ext_final$rsid[fuma_data_ext_final$Gene == "?"])
```

# Importing enhancer associations obtained from downloaded EnhancerAtlas 2.0 data
```{r MVP-ExtendedData-Final-v1.7-14}
lung_enh_gene <- read.table("Histone ChIP-Seq/lung_enhancer_gene.txt", sep="\t", header=TRUE)
liver_enh_gene <- read.table("Histone ChIP-Seq/liver_enhancer_gene.txt", sep="\t", header=TRUE)
cd34_enh_gene <- read.table("Histone ChIP-Seq/cd34_enhancer_gene.txt", sep="\t", header=TRUE)
spleen_enh_gene <- read.table("Histone ChIP-Seq/spleen_enhancer_gene.txt", sep="\t", header=TRUE)

lung_enh_gene$Tissue <- "Lung"
liver_enh_gene$Tissue <- "Liver"
cd34_enh_gene$Tissue <- "CD34+"
spleen_enh_gene$Tissue <- "Spleen"

all_enh_gene <- rbind(lung_enh_gene, liver_enh_gene, spleen_enh_gene, cd34_enh_gene)
```

# Filtering our SNPs to find those inside enhancer regions for complement genes
```{r MVP-ExtendedData-Final-v1.7-15}
#Filtering to enhancers for our candidate genes
complement_enh <- all_enh_gene %>% filter(Gene %in% genes_of_interest)

#Seeing which SNPs are physically located between any complement enhancers start/end positions
alt_enh_snps <- snps_sepsis_ext_grch37 %>% 
  filter(Gene %in% genes_of_interest) %>% 
  rowwise() %>% 
  filter(any(pos >= complement_enh$start & pos <= complement_enh$end)) %>% 
  group_by(Phecode, Race, Gene, rsid) %>% 
  slice(1)

# Convert to data.tables for manipulation
library(data.table)
setDT(snps_sepsis_ext_grch37)
snps_sepsis_ext_grch37[, pos := as.numeric(pos)]
setDT(complement_enh)

# Combine the gene/tissue with the rsIDs that are associated with that enhancer
temp_dt <- snps_sepsis_ext_grch37[complement_enh, on = .(pos >= start, pos <= end), .(rsid, pos, start, end, Gene, Tissue), nomatch=0]

#Combine and remove duplicates
names(temp_dt)[names(temp_dt)=="Gene"] <- "enh_gene"
names(temp_dt)[names(temp_dt)=="start"] <- "enh_start"
names(temp_dt)[names(temp_dt)=="end"] <- "enh_end"
names(temp_dt)[names(temp_dt)=="Tissue"] <- "enh_tissue"
temp_dt <- temp_dt %>% group_by(rsid) %>% summarize(enh_gene_all = paste(na.omit(unique(enh_gene)), collapse = ", "), enh_tissue_all = paste(na.omit(unique(enh_tissue)), collapse = ", "), across(everything(), ~first(., na.rm = TRUE)), .groups = 'drop')

alt_enh_snps_combined <- alt_enh_snps %>% 
  left_join(., (temp_dt %>% select(-pos)), by="rsid") %>% 
  group_by(Phecode, Race, Gene, rsid, enh_gene_all, enh_tissue_all) %>% 
  arrange(desc(!is.na(enh_gene_all))) %>% 
  slice(1) %>% 
  ungroup()
```

#Add enhancer data to FUMA
```{r MVP-ExtendedData-Final-v1.7-16}
alt_enh_snps_combined$pos <- as.numeric(alt_enh_snps_combined$pos)
snps_ingene_old <- read.table("snps_sepsis_ingenes.txt", header = TRUE, sep = "\t")
regonly_old <- read.table("regonly_old.txt", header = TRUE, sep = "\t")

#Add in the SNPs in enhancer regions as defined by EnhancerAtlas
fuma_altreceptors_enh <- left_join(fuma_data_ext_final, (alt_enh_snps_combined %>% select(rsid, enh_gene_all, enh_tissue_all, enh_gene, enh_start, enh_end)), by = "rsid") %>% filter(Gene %in% genes_of_interest | grepl(paste(genes_of_interest, collapse="|"), enh_gene_all) | eqtl_symbol %in% genes_of_interest | in_gene==1)
n_distinct(fuma_altreceptors_enh$rsid)

#See if we have the same SNPs from before - we do
table(snps_ingene_old$rsid %in% fuma_altreceptors_enh$rsid)
```

# Convert back to GRCh38 build for CADD/RDB score assessment
```{r MVP-ExtendedData-Final-v1.7-17}
fuma_altreceptors_enh <- fuma_altreceptors_enh %>% left_join(., rsid_pos, by="rsid")
fuma_altreceptors_enh <- fuma_altreceptors_enh %>% mutate(pos_grch37 = pos)
fuma_altreceptors_enh <- fuma_altreceptors_enh %>% mutate(pos = pos_grch38) %>% select(-pos_grch38)
n_distinct(fuma_altreceptors_enh$rsid)
```

# Create a "get_CADD_score" function to replace missing data
```{r MVP-ExtendedData-Final-v1.7-18}
library(httr)
library(jsonlite)

# Function to get CADD scores with reverse base fallback
get_cadd_score <- function(chrom, pos, ref, alt) {
  # Construct the API URL
  base_url <- "https://cadd.gs.washington.edu/api/v1.0/GRCh38-v1.7/"
  url <- paste0(base_url, chrom, ":", pos, "_", ref, "_", alt)
  
  # Query the API
  response <- GET(url)
  
  if (status_code(response) == 200) {
    content <- content(response, "text", encoding = "UTF-8")
    data <- fromJSON(content)
    score <- data$PHRED  # Adjust field name based on the response structure
    
    # Handle multiple values
    if (length(score) > 1) {
        warning("Unexpected multiple values returned: ", paste(score, collapse = ", "), "\n")
        score <- round(max(as.numeric(score), na.rm = TRUE), 3)
    }    
    
    if (!is.null(data) && !is.null(score)) {
      return(score)
    } else {
      message("No valid CADD score for: ", chrom, ":", pos, "_", ref, "_", alt)
    }
  } else {
    message("API query failed with status: ", status_code(response), 
            " for ", chrom, ":", pos, "_", ref, "_", alt)
  }
  
  # If the result is NA, try reversing ref and alt
  url_reverse <- paste0(base_url, chrom, ":", pos, "_", alt, "_", ref)
  response_reverse <- GET(url_reverse)
  
  if (status_code(response_reverse) == 200) {
    content_reverse <- content(response_reverse, "text", encoding = "UTF-8")
    data_reverse <- fromJSON(content_reverse)
    score_reverse <- data_reverse$PHRED  # Adjust field name based on the response structure
    
   # Handle multiple values
    if (length(score_reverse) > 1) {
        warning("Unexpected multiple values returned: ", paste(score_reverse, collapse = ", "), "\n")
        score_reverse <- round(max(as.numeric(score_reverse), na.rm = TRUE), 3)
    }

    if (!is.null(data_reverse) && !is.null(score_reverse)) {
      return(score_reverse)
    } else {
      message("No valid CADD score for reversed bases: ", chrom, ":", pos, "_", alt, "_", ref)
    }
  } else {
    message("Reverse API query failed with status: ", status_code(response_reverse), 
            " for ", chrom, ":", pos, "_", alt, "_", ref)
  }
  
  # Return NA if both queries fail
  return(NA)
}
```

# Fixing missing CADD scores - downloading the missing scores
```{r MVP-ExtendedData-Final-v1.7-19}
# Apply our custom function from above to the missing CADD scores
missing_cadd_initial <- fuma_altreceptors_enh %>% filter(is.na(CADD) | CADD == "") %>% group_by(rsid, chrom, pos, ref, alt) %>% slice(1) %>% ungroup()

missing_cadd_initial$fixed_CADD <- mapply(get_cadd_score, missing_cadd_initial$chrom, missing_cadd_initial$pos, missing_cadd_initial$ref, missing_cadd_initial$alt)
```

# Fixing missing CADD scores - adding to the dataframe
```{r MVP-ExtendedData-Final-v1.7-20}
missing_cadd_initial <- missing_cadd_initial %>% select(rsid, fixed_CADD)
missing_cadd_initial$fixed_CADD <- as.numeric(missing_cadd_initial$fixed_CADD)

fuma_altreceptors_enh_withcadd <- fuma_altreceptors_enh %>% left_join(., missing_cadd_initial, by="rsid")

fuma_altreceptors_enh <- fuma_altreceptors_enh_withcadd %>% mutate(CADD = if_else(is.na(CADD) & !is.na(fixed_CADD), fixed_CADD, CADD))
```

# Fixing missing RDB scores - exporting for manual search
```{r MVP-ExtendedData-Final-v1.7-21}
setwd("/Users/kinman/Desktop/MVP_Data")

initial_missing_rdb_snps <- fuma_altreceptors_enh %>% filter(is.na(RDB) | RDB == "") %>% distinct(rsid)
write.table(initial_missing_rdb_snps$rsid, "initial_missing_rdb_snps.txt", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

# Replace missing RDB scores - importing scores from manual search
```{r MVP-ExtendedData-Final-v1.7-22}
initial_missing_rdb_scores <- read.table("initial_missing_rdb_scores.tsv", header = TRUE, sep = "\t") %>% select(rsids, ranking)

fuma_altreceptors_enh_withrdb <- fuma_altreceptors_enh %>% left_join(., initial_missing_rdb_scores, by=c("rsid"="rsids"))

#Manually add ones that are still missing
fuma_altreceptors_enh_withrdb$RDB[fuma_altreceptors_enh_withrdb$rsid == "rs3746037"] <- "1f"
fuma_altreceptors_enh_withrdb$RDB[fuma_altreceptors_enh_withrdb$rsid == "rs3095156"] <- "1f"
fuma_altreceptors_enh_withrdb$RDB[fuma_altreceptors_enh_withrdb$rsid == "rs12196237"] <- "1f"
fuma_altreceptors_enh_withrdb$RDB[fuma_altreceptors_enh_withrdb$rsid == "rs9501393"] <- "4"
fuma_altreceptors_enh_withrdb$RDB[fuma_altreceptors_enh_withrdb$rsid == "rs112839858"] <- "1f"

#Add back the "RDB_num" variable for filtering
fuma_altreceptors_enh <- fuma_altreceptors_enh_withrdb %>% mutate(RDB = if_else(is.na(RDB) & !is.na(ranking), ranking, RDB),
                                                    RDB_num = case_when(
                                                      RDB == "1a" | RDB == "1b" | RDB == "1c" | RDB == "1d" | RDB == "1e" | RDB == "1f" ~ "1",
                                                      RDB == "2a" | RDB == "2b" | RDB == "2c" ~ "2",
                                                      RDB == "3a" | RDB == "3b" ~ "3",
                                                      RDB == "4" ~ "4",
                                                      RDB == "5" ~ "5",
                                                      RDB == "6" ~ "6",
                                                      RDB == "7" ~ "7"
                                                    )) %>% select(-ranking)

#Ensure there are 0 missing values now
still_missing_rdb <- fuma_altreceptors_enh %>% filter(is.na(RDB)) %>% distinct(rsid)
n_distinct(fuma_altreceptors_enh$rsid)
```

# Get missing GTEX eQTL data
```{r MVP-ExtendedData-Final-v1.7-23}
library(httr)
library(jsonlite)
library(curl)
library(gtexr)

#Initiate the dataframes
gtex_data_initial <- data.frame() # No results on first pass
gtex_data_initial_reversed <- data.frame()

gtex_snps_initial <- fuma_altreceptors_enh %>% select(rsid, chrom, pos, ref, alt) %>% group_by(rsid) %>% slice(1)

#Reverse allele alignment -> forwards allele alignment produced 0 results
gtex_variantids_initial_reversed <- paste0("chr", gtex_snps_initial$chrom, "_", gtex_snps_initial$pos, "_", gtex_snps_initial$alt, "_", gtex_snps_initial$ref, "_b38")

#Using reversed ref/alt pairing which actually gives results
gtex_data_initial_reversed <- do.call(rbind, lapply(gtex_variantids_initial_reversed, function(x) {
  temp <- suppressMessages(gtexr::get_significant_single_tissue_eqtls(variantIds = x, datasetId = "gtex_v10", tissueSiteDetailIds = c("Lung", "Liver", "Whole_Blood", "Spleen")))
  rbind(gtex_data_initial_reversed, temp)
}))

#Filter to only what we need and only in the genes of interest
gtex_data_initial_reversed_filt <- na.omit(gtex_data_initial_reversed) %>% select(snpId, geneSymbolUpper, pValue, tissueSiteDetailId, nes) %>% filter(geneSymbolUpper %in% genes_of_interest) %>% rename("rsid"="snpId", "gtex_symbol"="geneSymbolUpper", "gtex_tissue"="tissueSiteDetailId", "NES"="nes", "gtex_pval"="pValue")

#Add an "effect allele" column
gtex_data_initial_reversed_filt <- gtex_data_initial_reversed_filt %>% left_join(., (fuma_altreceptors_enh %>% select(rsid, ref, alt)), by="rsid") %>% rowwise() %>% mutate(gtex_effectallele = ref) %>% select(-ref, -alt) %>% group_by(rsid, gtex_effectallele, gtex_tissue, gtex_symbol, gtex_pval, NES) %>% slice(1) %>% ungroup()

fuma_altreceptors_enh_withgtex <- fuma_altreceptors_enh %>% left_join(., gtex_data_initial_reversed_filt, by="rsid")
n_distinct(fuma_altreceptors_enh_withgtex$rsid)
```

# Exporting SNP list to obtain LDLink eQTL data
```{r MVP-ExtendedData-Final-v1.7-24}
ldlink_rsid_list <- unique(fuma_altreceptors_enh_withgtex$rsid)
write.table(ldlink_rsid_list, "~/Desktop/MVP_Data/ldlink_rsid_list.txt", quote = FALSE, col.names = FALSE, row.names = FALSE)
```

# Import saved LDLink data as a second source of missing eQTL data
```{r MVP-ExtendedData-Final-v1.7-25}
eqtl_LDlink_results <- read.table("eqtl_LDLink_results.txt", header = TRUE, quote = "\"", sep = "\t", na.strings = c("NA", "", "."), stringsAsFactors = FALSE)

eqtl_LDlink_results_filt <- eqtl_LDlink_results %>% filter(!is.na(RS_ID))
eqtl_ldlink_results_summarized <- eqtl_LDlink_results_filt %>% group_by(Query, LDRace, Gene_Symbol, Tissue) %>% summarize(
  ldlink_associated_rsids = paste(unique(na.omit(RS_ID)), collapse = ", "),
  across(everything(), ~first(.[Query==RS_ID], na.rm = TRUE)),
  .groups = 'drop'
) %>% select(-RS_ID)

fuma_altreceptors_enh <- fuma_altreceptors_enh_withgtex %>% mutate(LDRace = case_when(Race == "AFR" ~ "AFR",
                                                                   Race == "EUR" ~ "EUR",
                                                                   Race == "HIS" ~ "AMR",
                                                                   Race == "META" ~ "ALL"))

fuma_altreceptors_enh_withldlink <- fuma_altreceptors_enh %>% left_join(., eqtl_ldlink_results_summarized, by=c("rsid"="Query", "LDRace"))

n_distinct(fuma_altreceptors_enh_withldlink$rsid)

fuma_altreceptors_enh <- fuma_altreceptors_enh_withldlink %>% rename("ldlink_symbol"="Gene_Symbol", "ldlink_pval"="P_value", "ldlink_effect"="Effect_Size", "ldlink_tissue"="Tissue") %>% rowwise() %>% mutate(
  ldlink_effectallele = str_split(Effect_Allele_Freq, "=")[[1]][1],
  ldlink_noneffectallele = str_split(Non_effect_Allele_Freq, "=")[[1]][1])

table(with(fuma_altreceptors_enh, ldlink_effectallele == ref | ldlink_effectallele == alt))
table(with(fuma_altreceptors_enh, ldlink_noneffectallele == ref | ldlink_noneffectallele == alt))

#Correct some of the effect and non-effect alleles to match the format of the MVP dataset
fuma_altreceptors_enh$ldlink_effectallele[fuma_altreceptors_enh$rsid == "rs138470231"] <- "TC"
fuma_altreceptors_enh$ldlink_noneffectallele[fuma_altreceptors_enh$rsid == "rs138470231"] <- "T"
fuma_altreceptors_enh$ldlink_effectallele[fuma_altreceptors_enh$rsid == "rs112839858"] <- "CCT"
fuma_altreceptors_enh$ldlink_noneffectallele[fuma_altreceptors_enh$rsid == "rs112839858"] <- "C"
n_distinct(fuma_altreceptors_enh$rsid)

#Making sure they all match either the ref or alt alleles now
table(with(fuma_altreceptors_enh, ldlink_effectallele == ref | ldlink_effectallele == alt))
table(with(fuma_altreceptors_enh, ldlink_noneffectallele == ref | ldlink_noneffectallele == alt))
```

# Clean up fuma_altreceptors_enh dataframe
```{r MVP-ExtendedData-Final-v1.7-26}
fuma_altreceptors_enh <- fuma_altreceptors_enh %>% group_by(Phecode, Race, Gene, rsid, enh_gene_all, enh_tissue_all, alignedDirection, eqtl_symbol, tissue, gtex_symbol, gtex_tissue, ldlink_symbol, ldlink_tissue) %>% arrange(desc(!is.na(enh_gene_all))) %>% slice(1) %>% ungroup()
```

# Pull out the SNPs that are most likely to be impactful on our genes of interest
```{r MVP-ExtendedData-Final-v1.7-27}
#Pull out the in-gene SNPs
snps_ingene <- fuma_altreceptors_enh %>% filter(in_gene==1)
n_distinct(snps_ingene$rsid)

#Pull out the other potentially clinically meaningful SNPs
#CADD >12.37
cadd_snps <- fuma_altreceptors_enh %>% filter(CADD > 12.37) %>% group_by(Phecode, Race, Gene, rsid, enh_gene_all, enh_tissue_all, alignedDirection, eqtl_symbol, tissue, gtex_symbol, gtex_tissue, ldlink_symbol, ldlink_tissue) %>% slice(1) %>% ungroup() %>% arrange(rsid)
cadd_snplist <- unique(cadd_snps$rsid)
n_distinct(cadd_snps$rsid)

#RDB 3b or lower
reg_snps <- fuma_altreceptors_enh %>% filter(RDB_num < 4) %>% group_by(Phecode, Race, Gene, rsid, enh_gene_all, enh_tissue_all, alignedDirection, eqtl_symbol, tissue, gtex_symbol, gtex_tissue, ldlink_symbol, ldlink_tissue) %>% slice(1) %>% ungroup() %>% arrange(rsid)
reg_snplist <- unique(reg_snps$rsid)
n_distinct(reg_snps$rsid)

#eQTL data for AP genes
eqtl_snps <- fuma_altreceptors_enh %>% filter(eqtl_symbol %in% genes_of_interest | ldlink_symbol %in% genes_of_interest | gtex_symbol %in% genes_of_interest) %>% group_by(Phecode, Race, Gene, rsid, enh_gene_all, enh_tissue_all, alignedDirection, eqtl_symbol, tissue, gtex_symbol, gtex_tissue, ldlink_symbol, ldlink_tissue) %>% slice(1) %>% ungroup() %>% arrange(rsid)
eqtl_snplist <- unique(eqtl_snps$rsid)
n_distinct(eqtl_snps$rsid)

#Combine them all into one dataframe and remove any duplicates
snps_combined_fuma_data_ext <- rbind(snps_ingene, cadd_snps, reg_snps, eqtl_snps) %>% 
  group_by(Phecode, Race, Gene, rsid, enh_gene_all, enh_tissue_all, alignedDirection, eqtl_symbol, tissue, gtex_symbol, gtex_tissue, ldlink_symbol, ldlink_tissue) %>% slice(1) %>% ungroup()
n_distinct(snps_combined_fuma_data_ext$rsid)
n_distinct(snps_combined_fuma_data_ext$rsid[snps_combined_fuma_data_ext$in_gene==0])

#One last check to make sure all the SNPs are included - they are
table(snps_ingene_old$rsid %in% snps_combined_fuma_data_ext$rsid)
table(regonly_old$rsid %in% snps_combined_fuma_data_ext$rsid)
```

# Relabel the SNPs located inside any complement genes as "in_gene"
```{r MVP-ExtendedData-Final-v1.7-28}
# Label the in-gene SNPs by the gene they're located within
snps_combined_fuma_data_ext$chrom <- as.character(snps_combined_fuma_data_ext$chrom)

snps_combined_fuma_data_ext <- snps_combined_fuma_data_ext %>% 
  rowwise() %>%
  mutate(
    Gene = case_when(
      any(pos >= genes_by_loci$Start & pos <= genes_by_loci$End) ~ 
        genes_by_loci$Gene[which(pos >= genes_by_loci$Start & pos <= genes_by_loci$End)][1],
      TRUE ~ Gene
    ),
    in_gene = case_when(any(pos >= genes_by_loci$Start & pos <= genes_by_loci$End) ~ 1,
                        TRUE ~ 0)) %>% 
  ungroup()

#Create a separate dataframe to double-check them
snps_ingene_updated <- snps_combined_fuma_data_ext %>% filter(in_gene == 1)
snps_reg_updated <- snps_combined_fuma_data_ext %>% filter(in_gene == 0)
```

# Remove duplicates by grouping/slicing and update gene label based on enhancer/FUMA/positional data -> prioritize META datapoints by sorting races in reverse alphabetical order before slicing
```{r MVP-ExtendedData-Final-v1.7-29}
#Remove duplicates
snps_genereg_sliced <- snps_combined_fuma_data_ext %>% group_by(Phecode, Race, rsid, Gene, enh_gene_all, enh_tissue_all, alignedDirection, eqtl_symbol, tissue, gtex_symbol, gtex_tissue, ldlink_symbol, ldlink_tissue) %>% arrange(desc(Race), desc(!is.na(enh_gene_all))) %>% slice(1) %>% ungroup() %>% arrange(rsid)

#Once again label the genes using enhancer regions followed by ANNOVAR as preferred approach. If there's no ANNOVAR data, label based on the nearest gene or VEP gene. If none available, label with "?"
snps_genereg_labeled <- snps_genereg_sliced %>%
  mutate(
    Gene = case_when(
      !is.na(enh_gene_all) ~ enh_gene_all,
      !is.na(anno_symbol) ~ anno_symbol,
      !is.na(nearestGene) ~ nearestGene,
      !is.na(VEP_Gene) ~ VEP_Gene, 
      TRUE ~ "?"
    )
  ) %>%
  group_by(Phecode, Race, Gene, rsid, enh_gene_all, enh_tissue_all, 
           eqtl_symbol, tissue, gtex_symbol, gtex_tissue, ldlink_symbol, ldlink_tissue) %>%
  slice(1) %>%
  ungroup()

n_distinct(snps_genereg_labeled$rsid)
```

# Relabeling complement groupings by gene and filtering out the ones that have been annotated in our initial FUMA pass as not associated with our genes of interest
```{r MVP-ExtendedData-Final-v1.7-30}
snps_genereg_labeled <- snps_genereg_labeled %>% mutate(complement_group = case_when(
  Gene=="CFD" | Gene=="CFB" | Gene=="CFP" | Gene=="CFI" | Gene=="CFH" | Gene=="C3" ~ "alternative",
  Gene=="C1QA" | Gene=="C1QB" | Gene=="C1QC" | Gene=="C1R" | Gene=="C1S" | Gene=="C2" | Gene=="C4A" | Gene=="C4B" | Gene=="C4BPA" | Gene=="C4BPB" ~ "classical",
  Gene=="MBL2" | Gene=="MASP1" | Gene=="MASP2" ~ "lectin",
  Gene=="CR1" | Gene=="CR2" | Gene=="C3AR1" | Gene=="C5AR1" | Gene=="C5AR2" ~ "receptor",
  Gene=="CD46" | Gene=="CD55" | Gene=="CD59" ~ "regulator",
  Gene=="C5" | Gene=="C6" | Gene=="C7" | Gene=="C8A" | Gene=="C8B" | Gene=="C8G" | Gene=="C9" ~ "terminal",
  Gene=="CFHR1" | Gene=="CFHR2" | Gene=="CFHR3" | Gene=="CFHR4" ~ "other"))

snps_genereg_filt <- snps_genereg_labeled %>% filter((eqtl_symbol %in% genes_of_interest & eqtl_symbol != "CCHCR1") | (gtex_symbol %in% genes_of_interest & gtex_symbol != "CCHCR1") | (ldlink_symbol %in% genes_of_interest & ldlink_symbol != "CCHCR1") | (grepl(paste(genes_of_interest, collapse = "|"), Gene) & Gene != "CCHCR1"))

n_distinct(snps_genereg_filt$rsid)
```

# Export list of rsids for use in 1000Genomes allele frequency analysis to find the major/minor alleles using downloaded raw data
```{r MVP-ExtendedData-Final-v1.7-31}
rsid_1000G_list_v2 <- snps_genereg_filt %>% ungroup() %>% distinct(rsid)

write.table(rsid_1000G_list_v2, "rsid_1000G_list_v2.txt", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

# Extracting 1000Genomes frequencies from raw data to label major/minor alleles
```{r MVP-ExtendedData-Final-v1.7-32}
#Import the 1000Genomes frequencies for our SNPs
freq_data <- read.csv("freq_data_genereg_v2.txt", header = TRUE, sep = "") %>% select(POS, ID, REF, ALT, SAMN10492705)

#Separate out the multiple minor alleles and counts
freq_data <- freq_data %>% mutate(other_1 = sapply(strsplit(ALT, ","), function(x) x[2]),
                                  other_2 = sapply(strsplit(ALT, ","), function(x) x[3]),
                                  alt_allele = sapply(strsplit(ALT, ","), function(x) x[1]),
                                  total_alleles = as.numeric(sapply(strsplit(SAMN10492705, ":"), function(x) x[1])),
                                  alt_allele_counts = sapply(strsplit(SAMN10492705, ":"), function(x) x[2]))

freq_data <- freq_data %>% mutate(alt_count = as.numeric(sapply(strsplit(alt_allele_counts, ","), function(x) x[1])),
                                  other_1_count = as.numeric(sapply(strsplit(alt_allele_counts, ","), function(x) x[2])),
                                  other_2_count = as.numeric(sapply(strsplit(alt_allele_counts, ","), function(x) x[3])))

#Remove NAs to enable logic to function then pick the most frequent minor allele as the true minor allele
freq_data <- freq_data %>% mutate(
    alt_count = if_else(is.na(alt_count), 0, alt_count),
  other_1_count = if_else(is.na(other_1_count), 0, other_1_count),
  other_2_count = if_else(is.na(other_2_count), 0, other_2_count),
  ALT = if_else(
    alt_count > other_1_count & alt_count > other_2_count | (alt_count==0 & other_1_count==0), alt_allele, if_else(
      other_1_count > other_2_count, other_1, other_2)),
  alt_count = as.numeric(if_else(
    alt_count > other_1_count & alt_count > other_2_count, alt_count, if_else(other_1_count > other_2_count, other_1_count, other_2_count))),
                                  MAF = alt_count/total_alleles)

freq_data <- freq_data %>%
  mutate(MAF = format(MAF, scientific = FALSE)) %>% arrange(ID)

freq_data_mismatches <- freq_data %>% filter((!(REF %in% snps_genereg_filt$ref) & !(REF %in% snps_genereg_filt$alt)) | (!(ALT %in% snps_genereg_filt$ref) & !(ALT %in% snps_genereg_filt$alt)))

#Correct alleles to match our MVP notation - otherwise this will not correctly assign the alleles
allele_check <- snps_genereg_filt %>% group_by(rsid) %>% slice(1) %>% arrange(rsid)

freq_data$REF[freq_data$ID=="rs59449659"] <- "CAA"
freq_data$REF[freq_data$ID=="rs368112020"] <- "TA"
freq_data$ALT[freq_data$ID=="rs368112020"] <- "T"
freq_data$REF[freq_data$ID=="rs147277810"] <- "C"
freq_data$REF[freq_data$ID=="rs531356471"] <- "CA"
freq_data$REF[freq_data$ID=="rs560054595"] <- "GC"
freq_data$ALT[freq_data$ID=="rs560054595"] <- "G"
freq_data$REF[freq_data$ID=="rs368253378"] <- "C"
freq_data$ALT[freq_data$ID=="rs368253378"] <- "CA"
freq_data$REF[freq_data$ID=="rs9279402"] <- "TA"

#Repeat to make sure there are no remaining mismatches
freq_data_mismatches <- freq_data %>% filter((!(REF %in% snps_genereg_filt$ref) & !(REF %in% snps_genereg_filt$alt)) | (!(ALT %in% snps_genereg_filt$ref) & !(ALT %in% snps_genereg_filt$alt)))

#Assign major and minor allele roles
freq_data <- freq_data %>% mutate(major_allele = if_else(MAF>0.50, ALT, REF),
                                  minor_allele = if_else(MAF>0.50, REF, ALT)) %>% arrange(ID)
```

# Standardize OR/beta based on 1000Genomes data so reference allele = major allele and alt allele = minor allele
```{r MVP-ExtendedData-Final-v1.7-33}
snps_genereg_alleles <- snps_genereg_filt %>% left_join(., (freq_data %>% select(ID, major_allele, minor_allele)), by=c("rsid"="ID"))

#Confirming that ref and alt alleles appropriately match with major or minor alleles so that the odds ratio correction works appropriately - rs11808491 is the only mismatch (T>C ref/alt, C>A major/minor)
table(with(snps_genereg_alleles, ref==major_allele | ref==minor_allele))
table(with(snps_genereg_alleles, alt==major_allele | alt==minor_allele))

#rs11808491 should be C/T based on 1000Genomes data, changing minor allele
snps_genereg_alleles$minor_allele[snps_genereg_alleles$rsid=="rs11808491"] <- "T"

#Check again to make sure they all match now - they do
table(with(snps_genereg_alleles, ref==major_allele | ref==minor_allele))
table(with(snps_genereg_alleles, alt==major_allele | alt==minor_allele))

#If reference allele is the major allele, leave the OR alone - otherwise flip it and take the inverse
snps_genereg_corrected <- snps_genereg_alleles %>% mutate(or = if_else(ref == major_allele, or, 1/or),
                                                          beta = if_else(ref == major_allele, beta, (beta*-1)))

#Switch all columns so the reference allele is the major allele and the alt allele is the minor allele
snps_genereg_finalalleles <- snps_genereg_corrected %>% mutate(ref = major_allele,
                                                        alt = minor_allele)
```

# Set eQTL directions for each source of data based on major and minor alleles
```{r MVP-ExtendedData-Final-v1.7-34}
#If the risk-increasining allele is already the alt allele, then leave the direction alone -> otherwise swap
snps_genereg_finalalleles <- snps_genereg_finalalleles %>% mutate(eqtl_direction = case_when(
  RiskIncAllele == alt ~ alignedDirection,
  RiskIncAllele == ref ~ case_when(alignedDirection == "+" ~ "-",
                                   alignedDirection == "-" ~ "+"),
  RiskIncAllele != alt & RiskIncAllele != ref ~ "?"
))

#Same as above for the LDLink effect alleles
snps_genereg_finalalleles <- snps_genereg_finalalleles %>% mutate(
  ldlink_direction = if_else(ldlink_effectallele == alt,
    case_when(
      ldlink_effect <0 ~ "-",
      ldlink_effect >0 ~ "+"
    ),
    if_else(
      ldlink_effectallele == ref,
      case_when(
        ldlink_effect <0 ~ "+",
        ldlink_effect >0 ~ "-"
        ), "?"
      )
  )
)

#Same as above for the downloaded additional GTEX data - NES is the normalized effect size with +/- indicating direction of impact of effect on expression
snps_genereg_finalalleles <- snps_genereg_finalalleles %>% mutate(
  gtex_direction = if_else(gtex_effectallele == alt,
    case_when(
      NES <0 ~ "-",
      NES >0 ~ "+"
    ),
    if_else(
      gtex_effectallele == ref,
      case_when(
        NES <0 ~ "+",
        NES >0 ~ "-"
        ), "?"
      )
  )
)

#Collect any SNPs that could not be assigned and make sure there are none
eqtl_questions <- snps_genereg_finalalleles %>% filter(eqtl_direction == "?")
ldlink_questions <- snps_genereg_finalalleles %>% filter(ldlink_direction == "?")
gtex_questions <- snps_genereg_finalalleles %>% filter(gtex_direction == "?")

#See which directions don't match
overall_mismatches <- snps_genereg_finalalleles %>% filter(eqtl_direction != ldlink_direction | eqtl_direction != gtex_direction) %>% group_by(rsid, gtex_symbol, gtex_tissue, eqtl_symbol, tissue, ldlink_symbol, ldlink_tissue) %>% slice(1) %>% ungroup()

#Manually correct the mismatched values based on GTEX database (manually searching each SNP in their online database to confirm which allele is associated with which direction)
snps_genereg_finalalleles$eqtl_direction[snps_genereg_finalalleles$rsid == "rs1048709"] <- "+"
snps_genereg_finalalleles$eqtl_direction[snps_genereg_finalalleles$rsid == "rs71524421" & snps_genereg_finalalleles$eqtl_symbol == "CFHR3"] <- "+"
snps_genereg_finalalleles$eqtl_direction[snps_genereg_finalalleles$rsid == "rs71524421" & snps_genereg_finalalleles$gtex_symbol == "CFHR4"] <- "-"

#Repeat to ensure no mismatches now - there aren't any
overall_mismatches <- snps_genereg_finalalleles %>% filter(eqtl_direction != ldlink_direction | eqtl_direction != gtex_direction) %>% group_by(rsid, gtex_symbol, gtex_tissue, eqtl_symbol, tissue, ldlink_symbol, ldlink_tissue) %>% slice(1) %>% ungroup()
```

# Set up the eQTL and MVP data summary dataframe - we will use this to summarize eQTL data and prioritize META p-values, OR, and MAF
```{r MVP-ExtendedData-Final-v1.7-35}
#Make sure to preserve the pval, or, and maf from the META groups when it is present, otherwise take from the row with the highest number of cases (num_cases)
eqtl_races_genes <- snps_genereg_finalalleles %>%
  group_by(Phecode, Gene, rsid) %>%
  arrange(desc(Race == "META"), desc(num_cases)) %>%
  summarize(
    races_combined = paste(na.omit(unique(Race)), collapse = ", "),
    
    eqtl_genes_all = paste(
      unique(paste0(
        eqtl_symbol[!is.na(eqtl_symbol) & !is.na(eqtl_direction)], " (",
        eqtl_direction[!is.na(eqtl_symbol) & !is.na(eqtl_direction)], ")"
        )),
      collapse = ", "
      ),
    eqtl_genes_complement = paste(
      unique(paste0(
        eqtl_symbol[eqtl_symbol %in% genes_of_interest & !is.na(eqtl_symbol) & !is.na(eqtl_direction)], " (",
        eqtl_direction[eqtl_symbol %in% genes_of_interest & !is.na(eqtl_symbol) & !is.na(eqtl_direction)], ")"
        )),
      collapse = ", "
    ),
    
    ldlink_genes_all = paste(
      unique(paste0(
        ldlink_symbol[!is.na(ldlink_symbol) & !is.na(ldlink_direction)], " (",
        ldlink_direction[!is.na(ldlink_symbol) & !is.na(ldlink_direction)], ")"
        )),
      collapse = ", "
      ),
    
    gtex_genes_all = paste(
      unique(paste0(
        gtex_symbol[!is.na(gtex_symbol) & !is.na(gtex_direction)], " (",
        gtex_direction[!is.na(gtex_symbol) & !is.na(gtex_direction)], ")"
        )),
      collapse = ", "
      ),
    
    eqtl_tissues_complement = paste(
      na.omit(unique(tissue[eqtl_symbol %in% genes_of_interest])),
      collapse = ", "
    ),
    
    ldlink_tissues_all = paste(na.omit(unique(ldlink_tissue)), collapse = ", "),
    gtex_tissues_all = paste(na.omit(unique(gtex_tissue)), collapse = ", "),
    
    pval = first(pval),
    or = first(or),
    maf = first(maf),
    
    .groups = 'drop'
  ) %>%
  select(-Gene) %>%
  mutate(across(where(is.character), ~ gsub("\\(\\)", "", .))) %>% 
  mutate(across(where(is.character), ~ gsub("\\s*,\\s*,", ", ", .))) %>%
  mutate(across(where(is.character), ~ gsub("^\\s*,|,\\s*$", "", .))) %>%
  mutate(across(where(is.character), trimws))

#Combine the different eQTL data sources into one - use nchar to pick whichever one has the most data included (ie, the longest list of genes/tissues)
prioritize_source <- function(gtex, ldlink, original) {
  options <- c(gtex, ldlink, original)
  lengths <- nchar(options)
  priority <- which(lengths == max(lengths))[1]  # chooses first in priority order on tie
  options[priority]
}

eqtl_races_genes_combined <- eqtl_races_genes %>% 
  rowwise() %>% 
  mutate(
    eqtl_genes_complement = prioritize_source(gtex_genes_all, ldlink_genes_all, eqtl_genes_complement),
    eqtl_tissues_complement = prioritize_source(gtex_tissues_all, ldlink_tissues_all, eqtl_tissues_complement)
    ) %>% 
  select(Phecode, rsid, races_combined, maf, or, pval, ends_with("_complement"))

#Make Whole_Blood look prettier for table-making purposes
eqtl_races_genes_combined$eqtl_tissues_complement <- gsub("Whole_Blood", "Whole Blood", eqtl_races_genes_combined$eqtl_tissues_complement)
```

# Add the summarized eQTL data with META maf, OR, and p-values to get final condensed list of genic and potential regulatory SNPs
```{r MVP-ExtendedData-Final-v1.7-36}
snps_genereg_combined <- snps_genereg_finalalleles %>% select(-pval, -or, -maf) %>% left_join(., eqtl_races_genes_combined, by=c("rsid", "Phecode"))

#Condense everything with summarized complement eQTL data and previously-summarized enhancer data
fuma_genereg_combined <- snps_genereg_combined %>% group_by(Phecode, races_combined, rsid, eqtl_genes_complement, eqtl_direction, eqtl_tissues_complement, enh_gene_all, enh_tissue_all) %>% arrange(desc(!is.na(eqtl_genes_complement) & eqtl_genes_complement != ""), desc(!is.na(enh_gene_all)), desc(Race)) %>% slice(1) %>% ungroup() %>% arrange(rsid)
n_distinct(fuma_genereg_combined$rsid)

#Make sure there are no duplicates that we're missing - there should be 0
duplicates <- fuma_genereg_combined %>% filter(duplicated(fuma_genereg_combined$rsid)) %>% group_by(Phecode, rsid) %>% summarize(dup_count = n(), .groups = 'drop') %>% ungroup() %>% filter(dup_count >1)
```

# Compare the previously-identified genic SNPs from before we included regulatory elements to ensure everything is included - all correct
```{r MVP-ExtendedData-Final-v1.7-37}
snps_ingene_old <- read.table("snps_sepsis_ingenes.txt", header = TRUE, sep = "\t")

table(snps_ingene_old$rsid %in% fuma_genereg_combined$rsid)
```

# Summarize by eQTL data to get final condensed list of genic and potential regulatory SNPs
```{r MVP-ExtendedData-Final-v1.7-38}
#Condense everything with summarized complement eQTL data
fuma_genereg_final <- fuma_genereg_combined %>% 
  group_by(Phecode, races_combined, rsid, eqtl_genes_complement, eqtl_tissues_complement, enh_gene_all, enh_tissue_all) %>% 
  arrange(desc(!is.na(eqtl_genes_complement) & eqtl_genes_complement != ""), desc(Race)) %>% 
  slice(1) %>% 
  ungroup() %>% 
  arrange(rsid)
n_distinct(fuma_genereg_final$rsid)

#Make sure there are no duplicates that we're missing - 0 found
duplicates_final <- fuma_genereg_final %>% filter(duplicated(fuma_genereg_final$rsid)) %>% group_by(Phecode, rsid) %>% summarize(dup_count = n(), .groups = 'drop') %>% ungroup() %>% filter(dup_count >1)
```

# Making sure the racial groups are all combined appropriately - they are, no mismatches
```{r MVP-ExtendedData-Final-v1.7-39}
snps_sig_race <- snps_noindels_altsig_filt %>% filter((complement_group == "alternative" | complement_group == "receptor" | complement_group == "regulator" | complement_group == "other") & lj_sig==1 & condition %in% sepsis_conditions & maf>maf_cutoff) %>% group_by(Phecode, rsid) %>% summarize(races_combined = paste(na.omit(unique(Race)), collapse = ", "), .groups = 'drop')

fuma_genereg_final_test <- fuma_genereg_final %>% left_join(., snps_sig_race, by=c("Phecode", "rsid")) %>% mutate(races_combined.x = if_else(!is.na(races_combined.y) & races_combined.x != races_combined.y, races_combined.y, races_combined.x)) %>% rename("races_combined"="races_combined.x") %>% select(-races_combined.y)

fuma_genereg_test <- fuma_genereg_final_test %>% left_join(., snps_sig_race, by=c("Phecode", "rsid"))
racial_mismatches <- fuma_genereg_test %>% filter(races_combined.x != races_combined.y)
```

# Data QC and cleanup
```{r MVP-ExtendedData-Final-v1.7-40}
fuma_genereg_final$eqtl_tissues_complement <- gsub("Cells_EBV-transformed_lymphocytes, ", "", fuma_genereg_final$eqtl_tissues_complement)
fuma_genereg_final$eqtl_tissues_complement <- gsub(", Cells_EBV-transformed_lymphocytes", "", fuma_genereg_final$eqtl_tissues_complement)

#For the potential regulatory SNPs assigned to 2 different genes, take the one with the lowest distance to the gene based on ANNOVAR. Don't exclude samples if this value is NA though since then we lose some SNPs that are missing that data
dups <- fuma_genereg_final %>% group_by(Phecode, rsid) %>% summarize(dup_count = n(), .groups = 'drop') %>% filter(dup_count >1)
dup_data <- fuma_genereg_final %>% filter(rsid %in% dups$rsid)

fuma_genereg_final_collapsed <- fuma_genereg_final %>% group_by(rsid, races_combined, Phecode) %>% filter(dist_annov == min(dist_annov) | is.na(dist_annov)) %>% ungroup()
n_distinct(fuma_genereg_final_collapsed$rsid)

fuma_genereg_final_filt <- fuma_genereg_final_collapsed %>% filter((in_gene == 1 | grepl(paste(genes_of_interest, collapse = "|"), eqtl_genes_complement)) | (in_gene == 0 & (CADD >12.37 | RDB_num < 4)))
n_distinct(fuma_genereg_final_filt$rsid)

issues <- fuma_genereg_final_collapsed %>% filter(!(rsid %in% fuma_genereg_final_filt$rsid))
```

# Testing to see if this contains all of our prior SNPs - it does
```{r MVP-ExtendedData-Final-v1.7-41}
fuma_genereg_final_filt <- fuma_genereg_final_filt %>% arrange(rsid)
snps_ingene_old <- snps_ingene_old %>% arrange(rsid)

table(snps_ingene_old$rsid %in% fuma_genereg_final_filt$rsid)
table(snps_genereg_combined$rsid %in% fuma_genereg_final_filt$rsid)
table(with(fuma_genereg_final_filt, rsid %in% snps_genereg_combined$rsid | rsid %in% snps_ingene_old$rsid))
```

# Making sure the genic SNPs are appropriately marked as such
```{r MVP-ExtendedData-Final-v1.7-42}
test_ingene <- fuma_genereg_final_filt %>% filter(in_gene==1)
test_regonly <- fuma_genereg_final_filt %>% filter(in_gene==0)

cadd_regonly <- test_regonly %>% filter(CADD >12.37)
rdb_regonly <- test_regonly %>% filter(RDB_num <4)
eqtl_regonly <- test_regonly %>% filter(grepl(paste(genes_of_interest, collapse = "|"), eqtl_genes_complement))

paste("Unique in-gene SNPs:", n_distinct(test_ingene$rsid))
paste("Unique potential regulatory SNPs:", n_distinct(test_regonly$rsid))
paste("CADD >12.37:", n_distinct(cadd_regonly$rsid))
paste("RDB <4:", n_distinct(rdb_regonly$rsid))
paste("eQTL for AP genes:", n_distinct(eqtl_regonly$rsid))
```

# List the coding SNPs for labeling later
```{r MVP-ExtendedData-Final-v1.7-43}
coding_snps <- fuma_genereg_final_filt %>% filter(annot=="exonic") %>% select(rsid) %>% pull()
```

# Export for LD analysis
```{r MVP-ExtendedData-Final-v1.7-44}
fuma_genereg_simple <- fuma_genereg_final_filt %>% ungroup() %>% group_by(rsid) %>% arrange(desc(Race), desc(!is.na(RDB)), desc(!is.na(CADD)), desc(!is.na(enh_gene_all)), desc(!is.na(eqtl_genes_complement)), pval) %>% slice(1) %>% ungroup()

test <- fuma_genereg_simple %>% filter(rsid == "rs512559")

#Write table for export
write.table(fuma_genereg_simple, "fuma_genereg_simple.txt", sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)
```

# Add LD data to overall dataframe and then filter to only the lead SNPs
```{r MVP-ExtendedData-Final-v1.7-45}
#Import LD summary data
ld_summary <- read.csv("~/Desktop/MVP_Data/ld_summary_combined_v2.txt", sep="\t", header=TRUE)

#Get rid of the "secondary_rsid" field if it's the same as the dominant rsID so that there aren't duplicates listed in our footnotes later on
ld_summary_filt <- ld_summary %>% 
  select(-Gene, -group) %>% 
  mutate(secondary_rsid = if_else(secondary_rsid == dominant_rsid, NA, secondary_rsid))

#Summarize the secondary rsIDs into one comma-separated list
ld_summary_collapsed <- ld_summary_filt %>% group_by(dominant_rsid) %>% summarize(secondary_rsid_combined = paste(na.omit(unique(secondary_rsid)), collapse = ", "), .groups = 'drop')

fuma_genereg_final_withld <- fuma_genereg_final_filt %>% left_join(., ld_summary_collapsed, by=c("rsid"="dominant_rsid")) 

#Filter to only the lead SNPs
fuma_genereg_final_withld_filtered <- fuma_genereg_final_withld %>% filter(!(rsid %in% ld_summary_filt$secondary_rsid))
n_distinct(fuma_genereg_final_withld_filtered$rsid)
```

# Clean up data for tables by getting rid of the NAs
```{r MVP-ExtendedData-Final-v1.7-46}
fuma_genereg_final_withld_filtered$eqtl_genes_complement[fuma_genereg_final_withld_filtered$eqtl_genes_complement=="CCHCR1"] <- ""
fuma_genereg_final_withld_filtered$eqtl_genes_complement[is.na(fuma_genereg_final_withld_filtered$eqtl_genes_complement)] <- ""
fuma_genereg_final_withld_filtered$eqtl_direction[is.na(fuma_genereg_final_withld_filtered$eqtl_direction)] <- ""
fuma_genereg_final_withld_filtered$eqtl_tissues_complement[is.na(fuma_genereg_final_withld_filtered$eqtl_tissues_complement)] <- ""
fuma_genereg_final_withld_filtered$enh_tissue_all[is.na(fuma_genereg_final_withld_filtered$enh_tissue_all)] <- ""
fuma_genereg_final_withld_filtered$enh_gene_all[is.na(fuma_genereg_final_withld_filtered$enh_gene_all)] <- ""
```

# Add a column to identify situations where the lead SNP isn't significant in meta-analysis but secondary SNP is
```{r MVP-ExtendedData-Final-v1.7-47}
ld_summary_races <- ld_summary %>% left_join(., (eqtl_races_genes %>% select(Phecode, rsid, races_combined)), by=c("secondary_rsid"="rsid"))

ld_summary_races <- ld_summary_races %>% rowwise() %>% mutate(secondary_meta = case_when(grepl("META", races_combined) ~ 1, TRUE ~ 0)) %>% select(Phecode, Gene, group, dominant_rsid, secondary_rsid, secondary_meta)
ld_summary_races$secondary_meta <- as.numeric(ld_summary_races$secondary_meta)

#Summarize and create a list of secondary rsIDs again
ld_summary_races_collapsed <- ld_summary_races %>% group_by(Phecode, Gene, group, dominant_rsid) %>% summarize(secondary_rsid_combined = paste(unique(na.omit(secondary_rsid)), collapse = ", "),
          secondary_meta_combined = sum(secondary_meta),
          .groups = 'drop')

ld_summary_races_collapsed <- ld_summary_races_collapsed %>% mutate(secondary_meta = if_else(secondary_meta_combined >= 1, 1, 0))

fuma_genereg_final_withld_filtered_withmeta <- fuma_genereg_final_withld_filtered %>% left_join(., (ld_summary_races_collapsed %>% select(Phecode, dominant_rsid, secondary_meta)), by=c("rsid"="dominant_rsid","Phecode"))
```

# Add OR/beta CIs to final dataframe before splitting it up
```{r MVP-ExtendedData-Final-v1.7-48}
fuma_genereg_final_withld_filtered_withmeta <- fuma_genereg_final_withld_filtered_withmeta %>% rowwise() %>% 
  mutate(
    beta_lowci = beta - (sebeta * 1.96),
    beta_highci = beta + (sebeta * 1.96),
    OR_calc = exp(beta),
    OR_lowci = exp(beta - (1.96 * sebeta)),
    OR_highci = exp(beta + (1.96 * sebeta))
    ) %>% ungroup()
```

# Split up the dataframe into SNPs with in silico modeling data suggesting an impact and SNPs with eQTL data for AP proteins
```{r MVP-ExtendedData-Final-v1.7-49}
eqtl_ext <- fuma_genereg_final_withld_filtered_withmeta %>% filter(in_gene == 0 & (!is.na(eqtl_genes_complement) & eqtl_genes_complement != "" & lj_sig==1)) %>% arrange(rsid)
n_distinct(eqtl_ext$rsid)

genereg_only_withmeta <- fuma_genereg_final_withld_filtered_withmeta %>% filter(Gene %in% genes_of_interest)
n_distinct(genereg_only_withmeta$rsid)

genereg_only_filt <- genereg_only_withmeta %>% filter(in_gene == 1 | CADD >12.37 | RDB_num < 4) %>% group_by(rsid, races_combined, eqtl_direction, condition) %>% slice(1) %>% ungroup()

# Switch back to the prior method of reporting eQTL data
genereg_only_filt <- genereg_only_filt %>% 
  mutate(eqtl_direction_complement = str_extract(eqtl_genes_complement, "(?<=\\().*?(?=\\))"),
         eqtl_genes_complement = gsub("\\s*\\([^)]*\\)", "", eqtl_genes_complement, perl=TRUE))
         
n_distinct(genereg_only_filt$rsid)
```

# Check the total number of regulatory SNPs
```{r MVP-ExtendedData-Final-v1.7-50}
regonly_check_final <- genereg_only_filt %>% filter(in_gene == 0)
paste("Unique regulatory SNPs outside of genes:", n_distinct(regonly_check_final$rsid))
regonly_check_meta <- genereg_only_filt %>% filter((grepl("META", races_combined) | secondary_meta == 1) & in_gene == 0)
paste("Unique META regulatory SNPs outside of genes with in silico modeling data:", n_distinct(regonly_check_meta$rsid))

cadd_meta <- regonly_check_meta %>% filter(CADD >12.37)
rdb_meta <- regonly_check_meta %>% filter(RDB_num <4)
eqtl_meta <- regonly_check_meta %>% filter(grepl(paste(genes_of_interest, collapse = "|"), eqtl_genes_complement))

paste("CADD >12.37:", n_distinct(cadd_meta$rsid))
paste("RDB <4:", n_distinct(rdb_meta$rsid))
paste("AP eQTL data:", n_distinct(eqtl_meta$rsid))
```

# Filtering the eQTL data to only SNPs that are not reported elsewhere
```{r MVP-ExtendedData-Final-v1.7-51}
eqtl_ext_filt <- eqtl_ext %>% filter(!(rsid %in% genereg_only_filt$rsid))

paste("Unique regulatory SNPs outside of genes with AP eQTL data:", n_distinct(eqtl_ext_filt$rsid))
```

# Testing if this contains the prior SNPs
```{r MVP-ExtendedData-Final-v1.7-52}
setwd("/Users/kinman/Desktop/MVP_Data")

genereg_old <- read.csv("genereg_old.txt", header = TRUE, sep = "\t")
regonly_old <- read.csv("regonly_old.txt", header = TRUE, sep = "\t")

table(genereg_old$rsid %in% genereg_only_filt$rsid | genereg_old$rsid %in% eqtl_ext$rsid)
table(regonly_old$rsid %in% genereg_only_filt$rsid | regonly_old$rsid %in% eqtl_ext$rsid)

ingene_check <- genereg_only_filt %>% filter(in_gene == 1)
regonly_check <- genereg_only_filt %>% filter(in_gene == 0)
n_distinct(ingene_check$rsid)
n_distinct(regonly_check$rsid)
n_distinct(eqtl_ext$rsid)
```

# Exporting eQTL SNPs for LD analysis to filter eQTL SNPs to lead SNPs only
```{r MVP-ExtendedData-Final-v1.7-53}
write.table(eqtl_ext_filt, "~/Desktop/MVP_Data/eqtl_ld_snps.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

# Import completed LD analysis
```{r MVP-ExtendedData-Final-v1.7-54}
#Import LD summary data
eqtl_ld_summary <- read.csv("eqtl_ld_summary_combined_v2.txt", sep="\t", header=TRUE)

eqtl_ld_summary <- eqtl_ld_summary %>% select(-chrom, -group) %>% mutate(secondary_rsid = if_else(secondary_rsid == dominant_rsid, NA, secondary_rsid))
n_distinct(eqtl_ld_summary$secondary_rsid)
n_distinct(eqtl_ld_summary$dominant_rsid)

eqtl_ld_summary_collapsed <- eqtl_ld_summary %>% group_by(dominant_rsid) %>% summarize(secondary_rsid_combined = paste(na.omit(unique(secondary_rsid)), collapse = ", "), .groups = 'drop')
n_distinct(eqtl_ld_summary_collapsed$dominant_rsid)

eqtl_only_withld <- eqtl_ext_filt %>% select(-secondary_rsid_combined) %>% left_join(., eqtl_ld_summary_collapsed, by=c("rsid"="dominant_rsid")) 
n_distinct(eqtl_only_withld$rsid)

#Filter to only the dominant SNPs
eqtl_only_withld_filtered <- eqtl_only_withld %>% filter(!(rsid %in% eqtl_ld_summary$secondary_rsid))

n_distinct(eqtl_only_withld_filtered$rsid)
eqtl_dups <- eqtl_only_withld_filtered %>% filter(duplicated(rsid))
```

# Change dataframe setup for extragenic SNPs with eQTL data -> should be focused on the eQTL gene not the "annotated" gene
```{r MVP-ExtendedData-Final-v1.7-55}
library(stringr)
eqtl_expanded <- eqtl_only_withld_filtered %>% 
  separate_longer_delim(eqtl_tissues_complement, delim = ", ") %>% 
  separate_longer_delim(eqtl_genes_complement, delim = ", ") %>% 
  filter(!(eqtl_tissues_complement == "Cells_EBV-transformed_lymphocytes")) %>% 
  mutate(eqtl_direction_complement = str_extract(eqtl_genes_complement, "(?<=\\().*?(?=\\))"),
         eqtl_genes_complement = sub(" \\(.*\\)$", "", eqtl_genes_complement))

#Making it so the direction is based on whichever one increases risk for the listed condition
eqtl_expanded_dirchanges <- eqtl_expanded %>% mutate(eqtl_direction_complement = case_when(
  or > 1 ~ eqtl_direction_complement,
  or < 1 ~ case_when(
    eqtl_direction_complement == "+" ~ "-",
    eqtl_direction_complement == "-" ~ "+"
  )
)) %>% group_by(rsid, eqtl_genes_complement, condition, eqtl_tissues_complement, eqtl_direction_complement) %>% slice(1) %>% ungroup()

#Grouping by gene for alternative table format
eqtl_bygene_dirchanges <- eqtl_expanded_dirchanges %>% 
  group_by(condition, eqtl_genes_complement, eqtl_direction_complement) %>% 
  summarize(
    annotated_genes_all = paste(unique(na.omit(Gene)), collapse = ", "), 
    associated_rsids = paste(unique(na.omit(rsid)), collapse = ", "), 
    eqtl_tissues_complement = paste(unique(na.omit(eqtl_tissues_complement)), collapse = ", "),
    races_combined = races_combined[which.max(nchar(races_combined))],
    .groups = 'drop'
    )
```

# Creating a dataframe with eQTL data for extragenic SNPs in META population only
```{r MVP-ExtendedData-Final-v1.7-56}
eqtl_expanded_dirchanges_meta <- eqtl_expanded_dirchanges %>% filter(grepl("META", races_combined))

eqtl_bygene_dirchanges_meta <- eqtl_expanded_dirchanges_meta %>% 
  group_by(condition, eqtl_genes_complement, eqtl_direction_complement) %>% 
  summarize(
    annotated_genes_all = paste(unique(na.omit(Gene)), collapse = ", "), 
    associated_rsids = paste(unique(na.omit(rsid)), collapse = ", "), 
    eqtl_tissues_complement = paste(unique(na.omit(eqtl_tissues_complement)), collapse = ", "),
    races_combined = races_combined[which.max(nchar(races_combined))],
    .groups = 'drop'
    )
```

# Make another column for enhancer location for use in tables
```{r MVP-ExtendedData-Final-v1.7-57}
genereg_only_filt <- genereg_only_filt %>% mutate(enh_location = case_when(
  (annot == "intronic" | annot == "exonic") & enh_gene_all != "" ~ anno_symbol,
  annot == "intergenic" ~ "",
  TRUE ~ ""
))
```

# Export META genic SNPs for QC later
```{r MVP-ExtendedData-Final-v1.7-58}
meta_genic_snps <- genereg_only_filt %>% filter(grepl("META", races_combined) & in_gene==1) %>% distinct(rsid)
write.table(meta_genic_snps, "~/Desktop/MVP_Data/meta_genic_snps.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
n_distinct(meta_genic_snps$rsid)
```

# Remove the eqtl_direction field when no eqtl_genes_complement found
```{r}
genereg_only_filt <- genereg_only_filt %>% mutate(
  eqtl_direction = ifelse((is.na(eqtl_genes_complement) | eqtl_genes_complement == ""), "", eqtl_direction)
)
```

# Table 1: META genic SNPs associated with bacteremia
```{r MVP-ExtendedData-Final-v1.7-59}
library(gt)

#Pull the SNPs of interest for the first table
infection_conditions <- c("Septicemia", "Gram negative septicemia", "Gram positive septicemia", "Bacteremia")

infection_meta_snps_ingene <- genereg_only_filt %>% ungroup() %>% filter((condition %in% infection_conditions & in_gene==1) & (grepl("META", races_combined) | secondary_meta==1)) %>% arrange(condition, rsid, races_combined)
n_distinct(infection_meta_snps_ingene$rsid)

#Make the table
table1_infection_ingene_meta <- infection_meta_snps_ingene %>%  select(races_combined, Gene, rsid, condition, alt, maf, or, pval, RDB, CADD, eqtl_genes_complement, eqtl_direction, eqtl_tissues_complement) %>% group_by(condition) %>% 
  gt(rowname_col = "rsid") %>% 
  tab_spanner(
    label = "Annotation",
    columns = c(Gene,alt)
  ) %>% 
  tab_spanner(
    label = "MVP Meta-analysis Data",
    columns = c(or,maf,pval)
  ) %>% 
  tab_spanner(
    label = "Predictive Modeling",
    columns = c(RDB, CADD)
  ) %>% 
  tab_spanner(
    label = "eQTL Data (When Available)",
    columns = c(eqtl_genes_complement, eqtl_direction, eqtl_tissues_complement)
  ) %>% 
  cols_label(
    or = "OR",
    Gene = "Gene",
    alt = "Minor Allele",
    maf = "MAF",
    pval = "P-Value",
    eqtl_tissues_complement = "GTEX Tissues",
    eqtl_direction = "Direction",
    eqtl_genes_complement = "Affected Genes",
    races_combined = "Ancestral Groups"
  ) %>%
  tab_style(
    cells_column_labels(columns = everything()),
    style = cell_text(weight = "bold")
  ) %>% 
  fmt_number(
    columns = c(or,maf),
    decimals = 3
  ) %>%
  fmt_scientific(
    columns = c(pval),
    decimals = 2
  ) %>% 
  tab_header(
    title = md("**Table 1. Genic AP SNPs Associated with Bacteremia Phenotypes in MVP Meta-Analysis**")
  ) %>%
  tab_stubhead(
    label = md("**Phenotype**")
  ) %>% 
  cols_align(
    align = "center",
    columns = everything()
  ) %>%
  tab_style(
    style = cell_text(align = "center"), 
    locations = cells_stubhead()
  ) %>% 
  tab_options(
    table.width = pct(100)
  ) %>%
  tab_footnote(
    footnote = md("**Table 1: Genic AP SNPs associated with bacteremia phenotypes in MVP meta-analysis.** Phenotypes obtained from Phecode descriptions. All SNPs listed were obtained from MVP trans-ancestral meta-analyses (calculated via GWAMA) and met gene-specific significance thresholds obtained by Li and Ji transformation. Additional ancestral groups listed if this SNP was also significant upon individual ancestral group analysis. Exonic SNPs (synonymous and non-synonymous) are highlighted in bold. For significant SNPs in linkage disequilibrium (LD), only the \"lead\" SNP is listed here. Lead SNP selected based on p-value, functional impact, proximity to regulatory elements, and eQTL data.")
  ) %>% 
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_body(columns = everything(), rows = rsid %in% coding_snps)
  ) %>% 
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_stub(rows = rsid %in% coding_snps)
  ) %>%
  text_transform(
    locations = cells_body(
      columns = races_combined,
      rows = !(grepl("META", races_combined))
    ),
    fn = function(x) {
      paste0(x, "<sup>*</sup>")
    }
  )

#Add secondary rsid column as a footnote
for (i in seq_len(nrow(infection_meta_snps_ingene))) {
  if (!is.na(infection_meta_snps_ingene$secondary_rsid_combined[i]) &&
      infection_meta_snps_ingene$secondary_rsid_combined[i] != "") {
    
    table1_infection_ingene_meta <- table1_infection_ingene_meta %>%
      tab_footnote(
        footnote = paste0("Additional SNPs in LD: ", infection_meta_snps_ingene$secondary_rsid_combined[i]),
        locations = cells_stub(
          rows = i
        )
      )
  }
}

if(any(!(grepl("META", infection_meta_snps_ingene$races_combined)))) {
  table1_infection_ingene_meta <- table1_infection_ingene_meta %>%
    tab_footnote(
      footnote = md("<sup>*</sup>SNP is included here because it is in strong LD with another SNP that is significant in MVP meta-analysis, although the lead SNP is itself not significant in this population")
    )
}

#table1_infection_ingene_meta

gtsave(table1_infection_ingene_meta, "table1_infection_ingene_meta.html")
webshot(url = "table1_infection_ingene_meta.html", file = "table1_infection_ingene_meta.pdf")
```

# Table 2: META genic SNPs associated with sepsis
```{r MVP-ExtendedData-Final-v1.7-60}
library(gt)

#Pull the conditions and SNPs of interest
inflammation_conditions <- c("Sepsis and SIRS", "Sepsis", "Septic shock")

inflammation_meta_snps_ingene <- genereg_only_filt %>% ungroup() %>% filter((condition %in% inflammation_conditions & in_gene==1) & (grepl("META", races_combined) | secondary_meta==1)) %>% arrange(condition, rsid, races_combined, eqtl_genes_complement)

n_distinct(inflammation_meta_snps_ingene$rsid)
dups_table2 <- inflammation_meta_snps_ingene %>% filter(duplicated(rsid)) %>% pull(rsid)
dups_table2_data <- inflammation_meta_snps_ingene %>% filter(rsid %in% dups_table2)

inflammation_meta_snps_ingene_unique <- inflammation_meta_snps_ingene %>% group_by(rsid) %>% slice(1) %>% ungroup()
table(inflammation_meta_snps_ingene_unique$Gene)

#Make the table
table2_inflammation_ingene_meta <- inflammation_meta_snps_ingene %>%  select(races_combined, Gene, rsid, condition, alt, maf, or, pval, RDB, CADD, eqtl_genes_complement, eqtl_direction, eqtl_tissues_complement) %>% group_by(condition) %>% 
  gt(rowname_col = "rsid") %>% 
  tab_spanner(
    label = "Annotation",
    columns = c(Gene,alt)
  ) %>% 
  tab_spanner(
    label = "MVP Meta-analysis Data",
    columns = c(or,maf,pval)
  ) %>% 
  tab_spanner(
    label = "Predictive Modeling",
    columns = c(RDB, CADD)
  ) %>% 
  tab_spanner(
    label = "eQTL Data (When Available)",
    columns = c(eqtl_genes_complement, eqtl_direction, eqtl_tissues_complement)
  ) %>% 
  cols_label(
    or = "OR",
    Gene = "Gene",
    alt = "Minor Allele",
    maf = "MAF",
    pval = "P-Value",
    eqtl_tissues_complement = "GTEX Tissues",
    eqtl_direction = "Direction",
    eqtl_genes_complement = "Affected Complement Genes",
    races_combined = "Ancestral Groups"
  ) %>%
  tab_style(
    cells_column_labels(columns = everything()),
    style = cell_text(weight = "bold")
  ) %>% 
  fmt_number(
    columns = c(or,maf),
    decimals = 3
  ) %>%
  fmt_scientific(
    columns = c(pval),
    decimals = 2
  ) %>% 
  tab_header(
    title = md("**Table 2. Genic AP SNPs Associated with Sepsis Phenotypes in MVP Meta-Analysis**")
  ) %>%
  tab_stubhead(
    label = md("**Phenotype**")
  ) %>% 
  cols_align(
    align = "center",
    columns = everything()
  ) %>%
  tab_style(
    style = cell_text(align = "center"), 
    locations = cells_stubhead()
  ) %>% 
  tab_options(
    table.width = pct(100)
  ) %>%
  tab_footnote(
    footnote = md("**Table 2: Genic AP SNPs associated with sepsis phenotypes in MVP meta-analysis.** Phenotypes obtained from Phecode descriptions. All SNPs listed were obtained from MVP trans-ancestral meta-analyses (calculated via GWAMA) and met gene-specific significance thresholds obtained by Li and Ji transformation. Additional ancestral groups listed if this SNP was also significant upon individual ancestral group analysis. Exonic SNPs (synonymous and non-synonymous) are highlighted in bold. For significant SNPs in linkage disequilibrium (LD), only the \"lead\" SNP is listed here. Lead SNP selected based on p-value, functional impact, proximity to regulatory elements, and eQTL data.")
  ) %>% 
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_body(columns = everything(), rows = rsid %in% coding_snps)
  ) %>% 
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_stub(rows = rsid %in% coding_snps)
  ) %>%
  text_transform(
    locations = cells_body(
      columns = races_combined,
      rows = !(grepl("META", races_combined))
    ),
    fn = function(x) {
      paste0(x, "<sup>*</sup>")
    }
  )

#Add secondary rsid column as a footnote
for (i in seq_len(nrow(inflammation_meta_snps_ingene))) {
  if (!is.na(inflammation_meta_snps_ingene$secondary_rsid_combined[i]) &&
      inflammation_meta_snps_ingene$secondary_rsid_combined[i] != "") {
    
    table2_inflammation_ingene_meta <- table2_inflammation_ingene_meta %>%
      tab_footnote(
        footnote = paste0("Additional SNPs in LD: ", inflammation_meta_snps_ingene$secondary_rsid_combined[i]),
        locations = cells_stub(
          rows = i
        )
      )
  }
}

if(any(!(grepl("META", inflammation_meta_snps_ingene$races_combined)))) {
  table2_inflammation_ingene_meta <- table2_inflammation_ingene_meta %>%
    tab_footnote(
      footnote = md("<sup>*</sup>SNP is included here because it is in strong LD with another SNP that is significant in MVP meta-analysis, although the lead SNP is itself not significant in this population")
    )
}

#table2_inflammation_ingene_meta

gtsave(table2_inflammation_ingene_meta, "table2_inflammation_ingene_meta.html")
webshot(url = "table2_inflammation_ingene_meta.html", file = "table2_inflammation_ingene_meta.pdf")
```

# Table 3: META potential regulatory SNPs associated with bacteremia AND sepsis (COMBINED TABLE)
```{r MVP-ExtendedData-Final-v1.7-61}
library(gt)

combined_snps_meta_regonly <- genereg_only_filt %>% ungroup() %>% filter(condition %in% sepsis_conditions & maf > maf_cutoff & in_gene==0 & (grepl("META", races_combined) | secondary_meta == 1)) %>% arrange(desc(condition %in% infection_conditions), condition, rsid, races_combined)

n_distinct(combined_snps_meta_regonly$rsid)
combined_snps_meta_regonly_unique <- combined_snps_meta_regonly %>% group_by(rsid) %>% slice(1)
table(combined_snps_meta_regonly_unique$Gene)

n_distinct(combined_snps_meta_regonly$rsid)
dups_table3 <- combined_snps_meta_regonly %>% filter(duplicated(rsid)) %>% pull(rsid)
dups_table3_data <- combined_snps_meta_regonly %>% filter(rsid %in% dups_table3)

#Make the table
table3_combined_meta_regonly <- combined_snps_meta_regonly %>%  select(races_combined, Gene, rsid, condition, alt, maf, or, pval, RDB, CADD, enh_location, enh_tissue_all) %>% group_by(condition) %>% 
  gt(rowname_col = "rsid") %>% 
  tab_spanner(
    label = "Annotation",
    columns = c(Gene, alt)
  ) %>% 
  tab_spanner(
    label = "MVP Data",
    columns = c(or,maf,pval)
  ) %>% 
  tab_spanner(
    label = "Predictive Modeling",
    columns = c(RDB, CADD)
  ) %>% 
  tab_spanner(
    label = "Enhancer Info",
    columns = c(enh_location, enh_tissue_all)
  ) %>% 
  cols_label(
    alt = "Minor Allele",
    Gene = "Associated Gene",
    or = "OR",
    maf = "MAF",
    pval = "P-Value",
    races_combined = "Ancestral Groups",
    enh_location = "Location",
    enh_tissue_all = "Tissues Affected"
  ) %>%
  tab_style(
    cells_column_labels(columns = everything()),
    style = cell_text(weight = "bold")
  ) %>% 
  fmt_number(
    columns = c(or,maf),
    decimals = 3
  ) %>%
  fmt_scientific(
    columns = c(pval),
    decimals = 2
  ) %>% 
  tab_header(
    title = md("**Table 3. Potential Regulatory AP SNPs Associated with Bacteremia or Sepsis Phenotypes in MVP Meta-Analysis**")
  ) %>%
  tab_stubhead(
    label = md("**Phenotype**")
  ) %>% 
  cols_align(
    align = "center",
    columns = everything()
  ) %>%
  tab_style(
    style = cell_text(align = "center"), 
    locations = cells_stubhead()
  ) %>% 
  tab_options(
    table.width = pct(100)
  ) %>% 
  tab_footnote(
    footnote = md("**Table 3: Potential regulatory AP SNPs associated with bacteremia or sepsis phenotypes in MVP meta-analysis.** Phenotypes obtained from Phecode descriptions. For significant SNPs in linkage disequilibrium (LD), only the \"lead\" SNP is listed here. Lead SNP selected based on p-value, functional impact, proximity to regulatory elements, and eQTL data. All listed SNPs met gene-specific adjusted significance thresholds obtained by Li and Ji transformation in MVP trans-ancestral meta-analysis (obtained with GWAMA) or are in strong LD with a SNP that is significant in MVP trans-ancestral meta-analysis - the latter are marked with an asterisk if present. If a SNP met gene-specific significance thresholds in multiple ancestral groups, the additional ancestral groups will be listed here. P-value, odds ratios, and MAF for SNPs that are significant in multiple groups are based on MVP meta-analysis when available or the largest ancestral group based on number of cases reported when meta-analysis data are unavailable. Odds ratios are based on presence or absence of minor allele, which is treated as the risk allele. Gene association was determined by FUMA and ANNOVAR annotation or by localization within an AP gene enhancer region as defined by EnhancerAtlas. Enhancers are located within intronic regions of the genes listed in the \"Location\" column unless otherwise specified. Only SNPs with a high predicted functional impact by *in silico* modeling (CADD >12.37 or RDB 3b or lower) are included.")
  ) %>%
  text_transform(
    locations = cells_body(
      columns = races_combined,
      rows = !(grepl("META", races_combined))
    ),
    fn = function(x) {
      paste0(x, "<sup>*</sup>")
    }
  )

#Add secondary rsid column as a footnote
for (i in seq_len(nrow(combined_snps_meta_regonly))) {
  if (!is.na(combined_snps_meta_regonly$secondary_rsid_combined[i]) &&
      combined_snps_meta_regonly$secondary_rsid_combined[i] != "") {
    
    table3_combined_meta_regonly <- table3_combined_meta_regonly %>%
      tab_footnote(
        footnote = paste0("Additional SNPs in LD: ", combined_snps_meta_regonly$secondary_rsid_combined[i]),
        locations = cells_stub(
          rows = i
        )
      )
  }
}

if(any(!(grepl("META", combined_snps_meta_regonly$races_combined)))) {
  table3_combined_meta_regonly <- table3_combined_meta_regonly %>%
    tab_footnote(
      footnote = md("<sup>*</sup>SNP is included here because it is in strong LD with another SNP that is significant in MVP meta-analysis, although the lead SNP is itself not significant in this population")
    )
}

#Print the table
#table3_combined_meta_regonly

#Save for later publication
gtsave(table3_combined_meta_regonly, "table3_combined_meta_regonly.html")
webshot(url = "table3_combined_meta_regonly.html", file = "table3_combined_meta_regonly.pdf", vwidth = 1200, vheight = 800)
```

# Table 4: META extended eQTL data
```{r MVP-ExtendedData-Final-v1.7-62}
library(gt)

combined_snps_eqtlonly_bygene_meta <- eqtl_bygene_dirchanges_meta %>% arrange(desc(condition %in% infection_conditions), condition, eqtl_genes_complement)

eqtl_count <- combined_snps_eqtlonly_bygene_meta %>% separate_longer_delim(associated_rsids, delim = ",")
n_distinct(eqtl_count$associated_rsids)

#Make the table
table4_combined_eqtlonly_bygene_meta <- combined_snps_eqtlonly_bygene_meta %>% select(condition, eqtl_genes_complement, eqtl_direction_complement, eqtl_tissues_complement, associated_rsids) %>% group_by(condition) %>% 
  gt(rowname_col = "eqtl_genes_complement") %>% 
  cols_label(
    associated_rsids = "Associated rsIDs",
    eqtl_tissues_complement = "GTEX Tissues",
    eqtl_direction_complement = "eQTL Direction",
    eqtl_genes_complement = "Affected Genes"
  ) %>%
  tab_style(
    cells_column_labels(columns = everything()),
    style = cell_text(weight = "bold")
  ) %>% 
  tab_header(
    title = md("**Table 4. Additional SNPs with AP eQTL Data That are Associated with Sepsis or Bacteremia Phenotypes in MVP Meta-Analysis**")
  ) %>%
  tab_stubhead(
    label = md("**Phenotype and Affected Gene**")
  ) %>% 
  cols_align(
    align = "center",
    columns = everything()
  ) %>%
  tab_style(
    style = cell_text(align = "center"), 
    locations = cells_stubhead()
  ) %>% 
  tab_options(
    table.width = pct(100)
  ) %>% 
  cols_width(
    associated_rsids ~ pct(40),
    everything() ~ pct(15)
  ) %>%
  tab_footnote(
    footnote = md("**Table 4: Additional SNPs with AP eQTL data that are associated with sepsis or bacteremia phenotypes in MVP meta-analysis.** Phenotypes obtained from Phecode descriptions. eQTL data was obtained from GTEX and LDLink - the latter includes results from SNPs in strong LD with those listed here. All listed SNPs met adjusted significance thresholds obtained by Li and Ji transformation in trans-ancestral meta-analysis (obtained with GWAMA). Only SNPs beyond gene loci without significant *in silico* findings are included here. All listed SNPs are \"lead\" SNPs identified following LD clustering; secondary SNPs are not included in this list and are available in supplemental data. For eQTL data associated with genic SNPs, see Table 1. For SNPs with strong *in silico* predicted effects and eQTL data, see Table 2. Listed tissue types are based on the GTEX v8 database. Directionality was selected by association with an increased risk for the listed conditions.")
  )

#Print the table
#table4_combined_eqtlonly_bygene_meta

#Save for later publication
gtsave(table4_combined_eqtlonly_bygene_meta, "table4_combined_eqtlonly_bygene_meta.html")
webshot(url = "table4_combined_eqtlonly_bygene_meta.html", file = "table4_combined_eqtlonly_bygene_meta.pdf", vwidth = 1200, vheight = 800)
```

# SUPPLEMENTAL DATA 
# Export SNP list for all racial groups
```{r MVP-ExtendedData-Final-v1.7-63}
genic_snplist <- unique(snps_ingene_updated$rsid)
reg_snplist <- unique(snps_reg_updated$rsid)
combined_list <- data.frame(rsid = unique(unlist(list(genic_snplist, reg_snplist))))

#Exporting a list of genic snps to use the command line to pull every single p-value/OR for every race from the original unfiltered file (not imported here since it was too large)
write.table(genic_snplist, "genic_snplist.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(reg_snplist, "reg_snplist.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(combined_list, "combined_list.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

# Import SNP data for all races
```{r}
genic_snps_data <- read.table("genic_snps_data.txt", sep = "\t", header = TRUE, na.strings = 99)
reg_snps_data <- read.table("reg_snps_data.txt", sep = "\t", header = TRUE, na.strings = 99)

combined_snps_data <- full_join(genic_snps_data, reg_snps_data, relationship = "many-to-many")
```

# Create a function to get the MAF and major/minor alleles from 1000Genomes for each SNP
```{r}
get_maf <- function(rsid, pop = "ALFA:SAMN10492705") {
  backup_pop = "1000GENOMES:phase_3:ALL"
  backup_pop2 = "gnomADg:ALL"
  Sys.sleep(0.2)  # pause to avoid rate limiting
  
  url <- paste0("https://rest.ensembl.org/variation/human/", rsid, "?pops=1")
  
  res <- tryCatch(
    GET(url, add_headers("Content-Type" = "application/json")),
    error = function(e) return(NULL)
  )
  
  if (is.null(res) || status_code(res) != 200) return(NA)
  
  dat <- tryCatch(fromJSON(rawToChar(res$content)), error = function(e) return(NULL))
  if (is.null(dat)) return(NA)
  
  pops <- dat$populations
  if (!is.data.frame(pops)) return(NA)
  
  # Check population exists
  if (!"population" %in% colnames(pops)) return(NA)
  
  pop_row <- pops[pops$population == pop, , drop = FALSE]
  if (nrow(pop_row) <2) pop_row <- pops[pops$population == backup_pop, , drop = FALSE]
  if (nrow(pop_row) <2) pop_row <- pops[pops$population == backup_pop2, , drop = FALSE]
  
  if (nrow(pop_row) == 0 || !"frequency" %in% colnames(pop_row)) return(NA)
  
  # Sort rows by descending frequency
  sorted <- pop_row[order(-pop_row$frequency), ]
  
  maj <- sorted$allele[1]
  min <- if (nrow(sorted) >= 2) sorted$allele[2] else NA
  maf <- if (!is.na(min)) sorted$frequency[sorted$allele == min] else NA
  
  freq_data <- data.frame(
    rsid = rsid,
    major = maj,
    minor = min,
    maf = maf,
    stringsAsFactors = FALSE
  )
  return(freq_data)
}
```

# Get frequencies to correct alleles for supplemental data
```{r}
combined_snps_unique <- data.frame(rsid = unique(combined_snps_data$rsid))

#cat("Querying MAFs from Ensembl...\n")
#pb_maf_allraces <- txtProgressBar(min = 0, max = length(combined_snps_unique$rsid), style = 3)

# Use lapply to preserve data frame structure
#mafs_allraces_list <- lapply(seq_along(combined_snps_unique$rsid), function(i) {
#  setTxtProgressBar(pb_maf_allraces, i)
#  get_maf(combined_snps_unique$rsid[i])
#})

#close(pb_maf_allraces)

# Combine the list of data frames into a single data frame
library(dplyr)
mafs_allraces_list <- mafs_allraces_list[!is.na(mafs_allraces_list)]
mafs_allraces_df <- bind_rows(mafs_allraces_list)
```

# Correct the alleles
```{r}
combined_snps_withmaf <- combined_snps_data %>% 
  left_join(., mafs_allraces_df %>% rename("maf_ensembl"="maf"), by="rsid")

gene_correction <- fuma_genereg_final %>% select(rsid, Gene) %>% group_by(rsid) %>% slice(1) %>% ungroup()

# Manual correction to match MVP data
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs112839858"] <- "C"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs112839858"] <- "CCT"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs11393029"] <- "G"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs11393029"] <- "GT"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs138470231"] <- "T"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs138470231"] <- "TC"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs140342096"] <- "G"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs140342096"] <- "GC"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs141533269"] <- "AG"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs141533269"] <- "A"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs141655345"] <- "AT"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs141655345"] <- "A"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs143410348"] <- "T"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs143410348"] <- "TAA"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs150609676"] <- "A"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs150609676"] <- "AT"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs200348147"] <- "A"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs200348147"] <- "AG"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs34212723"] <- "T"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs34212723"] <- "TC"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs35015056"] <- "A"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs35015056"] <- "AT"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs35347398"] <- "GT"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs35347398"] <- "G"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs35966884"] <- "A"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs35966884"] <- "AG"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs367626689"] <- "CA"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs367626689"] <- "C"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs368112020"] <- "TA"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs368112020"] <- "T"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs368253378"] <- "C"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs368253378"] <- "CA"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs3842427"] <- "A"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs3842427"] <- "AC"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs4151658"] <- "C"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs4151658"] <- "CAT"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs547154491"] <- "CT"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs547154491"] <- "C"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs549577847"] <- "G"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs549577847"] <- "GT"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs568178832"] <- "ACC"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs568178832"] <- "A"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs59449659"] <- "C"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs59449659"] <- "CAA"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs60527232"] <- "AG"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs60527232"] <- "A"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs66788159"] <- "TA"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs66788159"] <- "T"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs67500966"] <- "T"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs67500966"] <- "TC"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs72458772"] <- "ACT"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs72458772"] <- "A"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs72576381"] <- "CT"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs72576381"] <- "C"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs74186579"] <- "TA"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs74186579"] <- "T"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs9332721"] <- "TGTA"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs9332721"] <- "T"

#Additional edits
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs1050685"] <- "T"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs1050685"] <- "C"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs13090575"] <- "G"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs13090575"] <- "A"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs432450"] <- "C"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs432450"] <- "G"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs568178460"] <- "C"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs568178460"] <- "A"
combined_snps_withmaf$major[combined_snps_withmaf$rsid == "rs1309036017"] <- "A"
combined_snps_withmaf$minor[combined_snps_withmaf$rsid == "rs1309036017"] <- "G"

#Frequency check to ensure no more mismatches or missing allele assignments
freq_check <- combined_snps_withmaf %>% 
  select(rsid, ref, alt, major, minor, maf, maf_ensembl) %>% 
  group_by(rsid) %>% 
  slice(1) %>% 
  ungroup() %>% 
  arrange(rsid)

mismatches <- freq_check %>% filter((alt != major & alt != minor) | (ref != major & ref != minor))
missing_freq <- freq_check %>% filter(is.na(major) | is.na(minor))
```

# Correct the odds ratios, maf, and beta now
```{r}
#If reference allele is the major allele, leave the OR alone - otherwise flip it and take the inverse
combined_snps_corrected <- combined_snps_withmaf %>% mutate(or = if_else(alt == major, 1/or, or),
                                                          beta = if_else(alt == major, (beta*-1), beta)) %>% 
  left_join(., (gene_correction %>% rename("Gene_corrected"="Gene")), by="rsid") %>% 
  left_join(., (lj_results %>% select("sig_threshold_bon","Gene")), by="Gene") %>% 
  rowwise() %>% 
  mutate(lj_sig = ifelse(pval < sig_threshold_bon, 1, 0),
         Gene = ifelse(!is.na(Gene_corrected), Gene_corrected, Gene)) %>% 
  select(-Gene_corrected) %>% 
  ungroup()

#Switch all columns so the reference allele is the major allele and the alt allele is the minor allele and then re-label the genes
combined_snps_corrected$chrom <- as.character(combined_snps_corrected$chrom)
genes_by_loci$Chromosome <- as.character(genes_by_loci$Chromosome)

combined_snps_labeled <- combined_snps_corrected %>% mutate(ref = major, alt = minor) %>% 
  rowwise() %>% 
  mutate(
    Gene = case_when(
      any(pos >= genes_by_loci$Start & 
            pos <= genes_by_loci$End & 
            chrom == genes_by_loci$Chromosome) ~ 
        genes_by_loci$Gene[which(pos >= genes_by_loci$Start & 
                                   pos <= genes_by_loci$End & 
                                   chrom == genes_by_loci$Chromosome)][1],
      TRUE ~ Gene
    ),
    in_gene = case_when(
      any(pos >= genes_by_loci$Start & 
            pos <= genes_by_loci$End &
            chrom == genes_by_loci$Chromosome &
            Gene == genes_by_loci$Gene) ~ 1,
      TRUE ~ 0
    )
  ) %>%
  ungroup()

#Simplify and remove duplicates
combined_snps_finalalleles <- combined_snps_labeled %>% 
  select(-maf_ensembl) %>% 
  group_by(Phecode, Race, Gene, rsid) %>% 
  slice(1) %>% 
  ungroup()
```

# Create wide format
```{r}
wide_snps <- combined_snps_finalalleles %>% pivot_wider(names_from = Race, values_from = c(or, pval, maf), names_glue = "{Race}_{.value}")

wide_snps_combined <- wide_snps %>%
  group_by(Phecode, Gene, rsid) %>% 
  dplyr::summarise(across(c("ID", "pos", "chrom", "ref", "alt", "beta", "sebeta"), ~first(., na.rm=TRUE)), across(c("AFR_or","AFR_pval","AFR_maf","EUR_or","EUR_pval","EUR_maf","HIS_or","HIS_pval","HIS_maf","META_or","META_pval", "META_maf", "lj_sig", "in_gene"), ~sum(., na.rm=TRUE)), .groups = 'drop')

#Switch back to binary since lj_sig and in_gene got summed -> any non-zero value will be switched to 1 for ease of tracking and documentation
wide_snps_combined$lj_sig <- ifelse(wide_snps_combined$lj_sig>0, 1, wide_snps_combined$lj_sig)
wide_snps_combined$in_gene <- ifelse(wide_snps_combined$in_gene>0, 1, wide_snps_combined$in_gene)

#Add back the complement groups and condition names
wide_snps_combined <- mutate(wide_snps_combined, complement_group=case_when(
  Gene=="CFD" | Gene=="CFB" | Gene=="CFP" | Gene=="CFI" | Gene=="CFH" | Gene=="C3" ~ "alternative",
  Gene=="C1QA" | Gene=="C1QB" | Gene=="C1QC" | Gene=="C1R" | Gene=="C1S" | Gene=="C2" | Gene=="C4A" | Gene=="C4B" | Gene=="C4BPA" | Gene=="C4BPB" ~ "classical",
  Gene=="MBL2" | Gene=="MASP1" | Gene=="MASP2" ~ "lectin",
  Gene=="CR1" | Gene=="CR2" | Gene=="C3AR1" | Gene=="C5AR1" | Gene=="C5AR2" | Gene=="C5AR1, C5AR2" ~ "receptor",
  Gene=="CD46" | Gene=="CD55" | Gene=="CD59" ~ "regulator",
  Gene=="C5" | Gene=="C6" | Gene=="C7" | Gene=="C8A" | Gene=="C8B" | Gene=="C8G" | Gene=="C9" ~ "terminal",
  Gene=="CFHR1" | Gene=="CFHR2" | Gene=="CFHR3" | Gene=="CFHR4" ~ "other"))

wide_snps_combined <- mutate(wide_snps_combined, condition=case_when(
  Phecode=="Phe_480" ~ "Pneumonia",
  Phecode=="Phe_480_1" ~ "Bacterial pneumonia",
  Phecode=="Phe_480_2" ~ "Viral pneumonia",
  Phecode=="Phe_480_3" ~ "Pneumonia due to fungus (mycoses)",
  Phecode=="Phe_480_5" ~ "Bronchopneumonia and lung abscess",
  Phecode=="Phe_480_11" ~ "Pneumococcal pneumonia",
  Phecode=="Phe_481" ~ "Influenza",
  Phecode=="Phe_483" ~ "Acute bronchitis and bronchiolitis",
  Phecode=="Phe_038" ~ "Septicemia",
  Phecode=="Phe_038_1" ~ "Gram negative septicemia",
  Phecode=="Phe_038_2" ~ "Gram positive septicemia",
  Phecode=="Phe_038_3" ~ "Bacteremia",
  Phecode=="Phe_117" ~ "Mycoses",
  Phecode=="Phe_994" ~ "Sepsis and SIRS",
  Phecode=="Phe_994_2" ~ "Sepsis",
  Phecode=="Phe_994_21" ~ "Septic shock"))

wide_snps_combined_filt <- wide_snps_combined %>% filter(lj_sig == 1 & condition %in% sepsis_conditions & grepl(paste(genes_of_interest, collapse = "|"), Gene))
```

# Adding in LD and making a text format to save the table
```{r}
wide_snps_withld <- wide_snps_combined_filt %>% left_join(., ld_summary_collapsed, by=c("rsid"="dominant_rsid"))

wide_snps_withld_filtered <- wide_snps_withld %>% filter(!(rsid %in% ld_summary_filt$secondary_rsid)) %>%
  mutate(across(c(AFR_pval, EUR_pval, HIS_pval, META_pval), as.numeric)) %>% 
  mutate(across(c(ends_with("_pval"), ends_with("_or")), 
                ~ ifelse(. == 0 | is.na(.), NA_real_, .)))

tables1_textformat <- wide_snps_withld_filtered %>%
  mutate(across(c(ends_with("_or"), ends_with("_maf")), ~round(., 4)))
```

# Table S1: all genic SNPs across all races
```{r MVP-ExtendedData-Final-v1.7-64}
wide_genic_snps_withld_filtered <- wide_snps_withld_filtered %>% filter(in_gene == 1)

library(gt)
tables1_genic_snps_all <- wide_genic_snps_withld_filtered %>%  dplyr::select(Gene,alt,rsid,condition,AFR_or,AFR_pval,EUR_or,EUR_pval,HIS_or,HIS_pval,META_or,META_pval) %>% group_by(condition) %>% 
  gt(rowname_col = "rsid") %>% 
  tab_spanner(
    label = "Annotation",
    columns=c(Gene,alt)
  ) %>% 
  tab_spanner(
    label="AFR",
    columns=c(AFR_or,AFR_pval)
  ) %>% 
  tab_spanner(
    label="EUR",
    columns=c(EUR_or,EUR_pval)
  ) %>% 
  tab_spanner(
    label="HIS",
    columns=c(HIS_or,HIS_pval)
  ) %>% 
  tab_spanner(
    label="META",
    columns=c(META_or,META_pval)
  ) %>% 
  cols_label(
    AFR_or = "OR",
    AFR_pval = "P-Value",
    EUR_or = "OR",
    EUR_pval = "P-Value",
    HIS_or = "OR",
    HIS_pval = "P-Value",
    META_or = "OR",
    META_pval = "P-Value",
    alt = "Minor Allele"
  ) %>%
  fmt_number(
    columns = c(AFR_or,EUR_or,HIS_or,META_or),
    decimals = 3
  ) %>%
  fmt_scientific(
    columns = c(AFR_pval,EUR_pval,HIS_pval,META_pval),
    decimals = 2,
  ) %>% 
  sub_missing(
    columns = everything(),
    missing_text = ""
  ) %>% 
  tab_header(
    title = md("**Table S1. Genic Alternative Pathway (AP) SNPs Associated with Sepsis and Bacteremia Phecodes Across All Races**")
  ) %>%
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_body(
      columns = AFR_pval,
      rows = AFR_pval < 0.05
    )
  ) %>%
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_body(
      columns = EUR_pval,
      rows = EUR_pval < 0.05
    )
  ) %>%
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_body(
      columns = HIS_pval,
      rows = HIS_pval < 0.05
    )
  ) %>%
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_body(
      columns = META_pval,
      rows = META_pval < 0.05
    )
  ) %>% 
  tab_footnote(
    footnote = md("**Table S1: Genic alternative pathway (AP) SNPs associated with sepsis and bacteremia phecodes across all races.** Phenotypes obtained from Phecode descriptions. All listed SNPs met gene-specific significance thresholds obtained by Li and Ji transformation in either individual ancestral groups  or trans-ancestral meta-analysis (obtained with GWAMA). P-values that met adjusted signifiance thresholds are highlighted in bold. Odds ratios are based on presence or absence of minor allele, which is treated as the risk allele. Gene assignment was determined by FUMA and ANNOVAR annotation.")
  )

#Add secondary rsid column as a footnote
for (i in seq_len(nrow(wide_genic_snps_withld_filtered))) {
  if (!is.na(wide_genic_snps_withld_filtered$secondary_rsid_combined[i]) &&
      wide_genic_snps_withld_filtered$secondary_rsid_combined[i] != "") {
    
    tables1_genic_snps_all <- tables1_genic_snps_all %>%
      tab_footnote(
        footnote = paste0("Additional SNPs in LD: ", wide_genic_snps_withld_filtered$secondary_rsid_combined[i]),
        locations = cells_stub(
          rows = i
        )
      )
  }
}

#tables1_genic_snps_all

gtsave(tables1_genic_snps_all, "tables1_genic_snps_all.html")
webshot(url = "tables1_genic_snps_all.html", file = "tables1_genic_snps_all.pdf", vwidth = 1200, vheight = 800)

write.table(tables1_textformat, "/Users/kinman/Desktop/MVP_Data/Supplementary\ Data/tables1_genic_snps_all.tsv", sep="\t", quote = FALSE, row.names = FALSE, col.names = TRUE)
```

# Table S2: all regulatory SNPs across all races
```{r MVP-ExtendedData-Final-v1.7-65}
regonly_snps_combined <- genereg_only_filt %>% ungroup() %>% filter(in_gene == 0 & maf > maf_cutoff & lj_sig == 1 & (CADD >12.37 | RDB_num <4)) %>% arrange(desc(condition %in% infection_conditions), condition)

library(gt)

#Make the table
tables2_regonly_all <- regonly_snps_combined %>% select(races_combined, Gene, rsid, condition, alt, maf, or, pval, RDB, CADD, enh_location, enh_tissue_all, annot) %>% group_by(condition) %>% 
  gt(rowname_col = "rsid") %>% 
  tab_spanner(
    label = "Annotation",
    columns = c(Gene, alt)
  ) %>% 
  tab_spanner(
    label = "MVP Data",
    columns = c(or,maf,pval)
  ) %>% 
  tab_spanner(
    label = "Predictive Modeling",
    columns = c(RDB, CADD)
  ) %>% 
  tab_spanner(
    label = "Enhancer Info",
    columns = c(enh_location, enh_tissue_all)
  ) %>% 
  cols_label(
    alt = "Minor Allele",
    Gene = "Associated Gene",
    or = "OR",
    maf = "MAF",
    pval = "P-Value",
    races_combined = "Ancestral Groups",
    enh_location = "Location",
    enh_tissue_all = "Tissues Affected"
  ) %>%
  tab_style(
    cells_column_labels(columns = everything()),
    style = cell_text(weight = "bold")
  ) %>% 
  fmt_number(
    columns = c(or,maf),
    decimals = 3
  ) %>%
  fmt_scientific(
    columns = c(pval),
    decimals = 2
  ) %>% 
  tab_header(
    title = md("**Table S2. Potential Regulatory Alternative Pathway (AP) SNPs by Predictive Modeling Associated with Sepsis or Bacteremia Phenotypes Across All Races**")
  ) %>%
  tab_stubhead(
    label = md("**Phenotype**")
  ) %>% 
  cols_align(
    align = "center",
    columns = everything()
  ) %>%
  tab_style(
    style = cell_text(align = "center"), 
    locations = cells_stubhead()
  ) %>% 
  tab_options(
    table.width = pct(100)
  ) %>% 
  tab_footnote(
    footnote = md("**Table S2: Potential regulatory alternative pathway (AP) SNPs by predictive modeling associated with sepsis or bacteremia phenotypes.** Phenotypes obtained from Phecode descriptions. All listed SNPs met gene-specific significance thresholds obtained by Li and Ji transformation in either individual ancestral groups or trans-ancestral meta-analysis (obtained with GWAMA). If a SNP met gene-specific significance thresholds in multiple ancestral groups, the additional ancestral groups will be listed here. P-value, odds ratios, and MAF for SNPs that are significant in multiple groups are based on MVP meta-analysis when available or the largest ancestral group based on number of cases reported when meta-analysis data are unavailable. Odds ratios are based on presence or absence of minor allele, which is treated as the risk allele. Gene association was determined by FUMA and ANNOVAR annotation or by localization within an AP gene enhancer region as defined by EnhancerAtlas. Enhancers are located within intronic regions of the genes listed in the \"Location\" column unless otherwise specified. Only SNPs with a high predicted functional impact by *in silico* modeling (CADD >12.37 or RDB 3b or lower) are included.")
  ) %>%
  text_transform(
    locations = cells_body(
      columns = enh_location,
      rows = (annot == "exonic")
    ),
    fn = function(x) {
      paste0(x, "<sup>*</sup>")
    }
  ) %>% 
  cols_hide(annot)

#Add secondary rsid column as a footnote
for (i in seq_len(nrow(regonly_snps_combined))) {
  if (!is.na(regonly_snps_combined$secondary_rsid_combined[i]) &&
      regonly_snps_combined$secondary_rsid_combined[i] != "") {
    tables2_regonly_all <- tables2_regonly_all %>%
      tab_footnote(
        footnote = paste0("Additional SNPs in LD: ", regonly_snps_combined$secondary_rsid_combined[i]),
        locations = cells_stub(
          rows = i
        )
      )
  }
}

if(any(regonly_snps_combined$annot == "exonic")) {
  tables2_regonly_all <- tables2_regonly_all %>%
    tab_footnote(
      footnote = md("<sup>*</sup>Enhancer is located within an exon in this gene")
    )
}

#Print the table
#tables2_regonly_all

#Save for later publication
gtsave(tables2_regonly_all, "tables2_regonly_all.html")
webshot(url = "tables2_regonly_all.html", file = "tables2_regonly_all.pdf", vwidth = 1200, vheight = 800)

write.table(regonly_snps_combined, "/Users/kinman/Desktop/MVP_Data/Supplementary\ Data/regonly_snps_combined.tsv", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)
```

# Table S3: all extended eQTL data across all races
```{r Table S3 - eQTL for all races}
library(gt)

combined_snps_eqtlonly_bygene <- eqtl_bygene_dirchanges %>% filter(condition %in% infection_conditions | condition %in% inflammation_conditions)

#Make the table
tables3_combined_eqtlonly_bygene <- combined_snps_eqtlonly_bygene %>% select(condition, eqtl_genes_complement, eqtl_direction_complement, eqtl_tissues_complement, associated_rsids) %>% group_by(condition) %>% 
  gt(rowname_col = "eqtl_genes_complement") %>% 
  cols_label(
    associated_rsids = "Associated rsIDs",
    eqtl_tissues_complement = "GTEX Tissues",
    eqtl_direction_complement = "Direction",
    eqtl_genes_complement = "Phenotype and Affected Gene"
  ) %>%
  tab_style(
    cells_column_labels(columns = everything()),
    style = cell_text(weight = "bold")
  ) %>% 
  tab_header(
    title = md("**Table S3. Additional SNPs with AP eQTL Data That are Associated with Sepsis or Bacteremia Phenotypes Across All Races**")
  ) %>%
  tab_stubhead(
    label = md("**Phenotype and Affected Gene**")
  ) %>% 
  cols_align(
    align = "center",
    columns = everything()
  ) %>%
  tab_style(
    style = cell_text(align = "center"), 
    locations = cells_stubhead()
  ) %>% 
  tab_options(
    table.width = pct(100)
  ) %>% 
  cols_width(
    associated_rsids ~ pct(40),
    everything() ~ pct(15)
  ) %>%
  tab_footnote(
    footnote = md("**Table 3: Additional SNPs with AP eQTL data that are associated with sepsis or bacteremia phenotypes across all races.** Phenotypes obtained from Phecode descriptions. All listed SNPs met adjusted significance thresholds obtained by Li and Ji transformation in either individual ancestral groups or trans-ancestral meta-analysis (obtained with GWAMA). Only SNPs beyond gene loci without significant *in silico* findings with eQTL data suggesting an impact on AP gene expression are included here. For eQTL data associated with genic SNPs across all ancestral groups, see Table S1. For SNPs with strong *in silico* predicted effects and eQTL data, see Table S2. Listed tissue types are based on the GTEX v8 database.")
  )

#Print the table
#tables3_combined_eqtlonly_bygene

#Save for later publication
gtsave(tables3_combined_eqtlonly_bygene, "tables3_combined_eqtlonly_bygene.html")
webshot(url = "tables3_combined_eqtlonly_bygene.html", file = "tables3_combined_eqtlonly_bygene.pdf", vwidth = 1200, vheight = 800)

write.table(combined_snps_eqtlonly_bygene, "/Users/kinman/Desktop/MVP_Data/Supplementary\ Data/eqtl_snps_combined.tsv", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)
```

# Table S5 (S4 is RegulomeDB score)
```{r}
tables5_ljresults <- lj_results %>% select(Gene, Chromosome, Meff, sig_threshold_bon) %>% filter(!(Gene == "CFP")) %>% 
  gt(rowname_col = "Gene") %>% 
  cols_label(
    Meff = "Effective Observations",
    sig_threshold_bon = "Adjusted Threshold",
  ) %>%
  tab_style(
    cells_column_labels(columns = everything()),
    style = cell_text(weight = "bold")
  ) %>% 
  tab_header(
    title = md("**Table S5. Adjusted Significance Thresholds by Gene**")
  ) %>%
  tab_stubhead(
    label = md("**Gene**")
  ) %>% 
  cols_align(
    align = "center",
    columns = everything()
  ) %>%
  tab_style(
    style = cell_text(align = "center"), 
    locations = cells_stubhead()
  ) %>% 
  fmt_number(
    columns = Meff,
    decimals = 2
  ) %>% 
  fmt_scientific(
    columns = sig_threshold_bon,
    decimals = 2
  ) %>% 
  tab_footnote(
    footnote = md("**Table S5: Adjusted significance thresholds by gene.** All calculations utilized the combined Phase 3 variant files (.vcf) for all populations (ALL) provided by the 1000Genomes project. Linkage disequilibrium data for each gene was obtained through .vcf processing with PLINK. The number of \"effective observations\" per gene was calculated by the Li and Ji analysis method in R. The \"adjusted threshold\" is the Boneferroni-corrected p-value using total number of effective observations for each gene. SNPs were only included in our analysis if their p-value was below this gene-specific p-value based on gene annotation assignments.")
  )

#tables5_ljresults

gtsave(tables5_ljresults, "tables5_ljresults.html")
webshot(url = "tables5_ljresults.html", file = "tables5_ljresults.pdf", vwidth = 1200, vheight = 800)
```

# Additional supplemental data
```{r Supplemental TSVs}
library(gt)

write.table(eqtl_only_withld_filtered, "/Users/kinman/Desktop/MVP_Data/Supplementary\ Data/eqtl_only_withld_filtered.tsv", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

eqtl_summary_table <- eqtl_ld_summary_collapsed %>% 
  gt()
gtsave(eqtl_summary_table, "eqtl_summary_table.html")
webshot(url = "eqtl_summary_table.html", file = "eqtl_summary_table.pdf", vwidth = 1200, vheight = 800)

write.table(fuma_genereg_final_withld_filtered_withmeta, "/Users/kinman/Desktop/MVP_Data/Supplementary\ Data/complete_ldfiltered_dataframe.tsv", sep = "\t", row.names = FALSE, quote = FALSE, col.names = TRUE)
```

# More QC
```{r Quality checking}
#Making sure MAF and pvalues are taken from META racial groups
maf_check <- (genereg_only_filt %>% filter(grepl("META", races_combined))) %>% left_join(., (snps_altsig_filt %>% filter(Race=="META") %>% rename("maf_meta"="maf", "pval_meta"="pval", "or_meta"="or") %>% select(Phecode, Race, rsid, maf_meta, or_meta, pval_meta)), by=c("Phecode", "rsid"))

#Confirming there are no mismatches (OR isn't checked because there are lots of mismatches due to switching of OR directions by allele as above)
maf_mismatch <- maf_check %>% filter(maf != maf_meta | pval != pval_meta)

#Checking to make sure the eQTL directions are okay
eqtl_expanded_check <- combined_snps_eqtlonly_bygene_meta %>% separate_longer_delim(associated_rsids, delim = ",")

#Manually searching each on GTEX or LDLink as needed
eqtl_check <- eqtl_expanded_dirchanges %>% filter(rsid %in% eqtl_expanded_check$associated_rsids) %>% select(rsid, condition, Gene, ref, alt, or, eqtl_genes_complement, eqtl_direction_complement, eqtl_tissues_complement, eqtl_direction, ldlink_direction, gtex_direction)
```

# Prepare data for forest plots
```{r MVP-ExtendedData-Final-v1.7-69}
genereg_forest <- genereg_only_filt %>% rowwise() %>% 
  mutate(
    beta_lowci = beta - (sebeta * 1.96),
    beta_highci = beta + (sebeta * 1.96),
    OR_calc = exp(beta),
    OR_lowci = exp(beta - (1.96 * sebeta)),
    OR_highci = exp(beta + (1.96 * sebeta))
    ) %>% ungroup()

eqtl_forest <- eqtl_expanded_dirchanges_meta %>% mutate(Gene = eqtl_genes_complement) %>% group_by(Phecode, rsid) %>% slice(1) %>% ungroup()

#All of them combined - 50 unique SNPs in meta-analysis
genereg_eqtl_forest <- full_join(genereg_forest, eqtl_forest) %>% group_by(Phecode, races_combined, rsid) %>% slice(1) %>% ungroup()
genereg_eqtl_forest_meta <- genereg_eqtl_forest %>% filter(grepl("META", races_combined))
n_distinct(genereg_eqtl_forest_meta$rsid)

#Without the eQTLs - 39 unique SNPs
genereg_forest_meta <- genereg_forest %>% filter(grepl("META", races_combined))
n_distinct(genereg_forest_meta$rsid)
```

# Figure 2 - forest plot of all META SNPs minus the eQTLs located in/around non-AP genes
```{r MVP-ExtendedData-Final-v1.7-70}
library(forestplot)

#Formatting data
genereg_meta_formatted_data <- genereg_forest_meta %>%
  mutate(infection_condition = if_else(condition %in% infection_conditions, 1, 0)) %>% 
  group_by(condition) %>%
  arrange(infection_condition, Gene, rsid, .by_group = TRUE) %>% 
  group_map(~ bind_rows(
    tibble(
      rsid = .y$condition,
      OR_calc = NA,
      OR_lowci = NA,
      OR_highci = NA,
      is_header = TRUE # Header row
    ),
    .x %>% mutate(
      rsid = paste0("      ", rsid), 
      is_header = FALSE
    )
  )) %>%
  map_dfr(~ tibble(infection_condition = .x$infection_condition[2], data = list(.x))) %>%
  arrange(desc(infection_condition)) %>%
  pull(data) %>% 
  bind_rows()
#Show infection conditions first

# Prepare data matrix
genereg_meta_table_text <- cbind(
  c("Variant/Condition", ifelse(grepl(paste(coding_snps, collapse = "|"), genereg_meta_formatted_data$rsid), paste0(genereg_meta_formatted_data$rsid, "*"), genereg_meta_formatted_data$rsid)),
  c("MAF", ifelse(genereg_meta_formatted_data$is_header==TRUE, "", round(genereg_meta_formatted_data$maf, 3))),
  c("Gene", ifelse(genereg_meta_formatted_data$is_header==TRUE, "", genereg_meta_formatted_data$Gene)),
  c("OR (95% CI)", sprintf(ifelse(!is.na(genereg_meta_formatted_data$OR_calc), "%.2f (%.2f-%.2f)", ""), genereg_meta_formatted_data$OR_calc, genereg_meta_formatted_data$OR_lowci, genereg_meta_formatted_data$OR_highci))
)

genereg_meta_fp <- forestplot(labeltext = genereg_meta_table_text,
           mean = c(NA, genereg_meta_formatted_data$OR_calc),
           lower = c(NA, genereg_meta_formatted_data$OR_lowci),
           upper = c(NA, genereg_meta_formatted_data$OR_highci),
           is.summary = c(TRUE, genereg_meta_formatted_data$is_header),
           zero = 1,
           xlog = FALSE,
           xticks = c(0.5, 1, 1.5, 2),
           boxsize = 0.2,
           col = fpColors(box = "royalblue", line = "black", summary = "royalblue"),
           title = "Associations of All Significant Meta-Analysis SNPs with Bacteremia or Sepsis Phenotypes",
           graph.pos = 4,
           shapes_gp = fpShapesGp(
             lines = gpar(lwd = 0.8),
             box = gpar(lwd = 1.5)
             ),
           txt_gp = fpTxtGp(
             label = gpar(cex = 0.7),
             ticks = gpar(cex = 0.7, lwd = 0.8)
             )
           ) %>%
  fp_set_style(
    summary = gpar(fontface = "bold"),
    align = if_else(genereg_meta_formatted_data$is_header==TRUE, "l", "c"),
  ) %>%
  fp_set_zebra_style("#EEEEEE")

tiff("~/Desktop/MVP_Data/CCE\ Files/Figure2_genereg_meta_forestplot.tiff", height=2400, width = 2400, res = 300)
genereg_meta_fp
dev.off()

pdf("~/Desktop/MVP_Data/genereg_meta_forestplot.pdf", width = 8, height = 10)
print(genereg_meta_fp)
dev.off()
```
